From: Ben Hutchings <ben@decadent.org.uk>
Date: Sat, 7 Nov 2009 20:21:37 +0000
Subject: [PATCH] USB: isight-firmware: declare MODULE_FIRMWARE

commit 4e0961d53020b0fed83168d469ce44e1b3f25898 upstream.

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---
 drivers/usb/misc/isight_firmware.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/drivers/usb/misc/isight_firmware.c b/drivers/usb/misc/isight_firmware.c
index b897f65..1a88e27 100644
--- a/drivers/usb/misc/isight_firmware.c
+++ b/drivers/usb/misc/isight_firmware.c
@@ -112,6 +112,8 @@ out:
 	return ret;
 }
 
+MODULE_FIRMWARE("isight.fw");
+
 static void isight_firmware_disconnect(struct usb_interface *intf)
 {
 }
-- 
1.6.6

