From: Ben Hutchings <ben@decadent.org.uk>
Date: Wed, 15 Jan 2014 02:01:08 +0000
Subject: ACPI: Avoid ABI change in 3.12.7
Forwarded: not-needed

Hide the new no_hotplug flag from genksyms.

--- a/include/acpi/acpi_bus.h
+++ b/include/acpi/acpi_bus.h
@@ -169,8 +169,12 @@ struct acpi_device_flags {
 	u32 ejectable:1;
 	u32 power_manageable:1;
 	u32 match_driver:1;
+#ifdef __GENKSYMS__
+	u32 reserved:27;
+#else
 	u32 no_hotplug:1;
 	u32 reserved:26;
+#endif
 };
 
 /* File System */
