--- a/arch/sh/Makefile	2009-12-03 12:51:21.000000000 +0900
+++ b/arch/sh/Makefile	2010-02-04 15:47:57.000000000 +0900
@@ -75,7 +75,6 @@
 
 # Give the various platforms the opportunity to set default image types
 defaultimage-$(CONFIG_SUPERH32)			:= zImage
-defaultimage-$(CONFIG_SH_SH7785LCR)		:= uImage
 defaultimage-$(CONFIG_SH_RSK)			:= uImage
 defaultimage-$(CONFIG_SH_URQUELL)		:= uImage
 defaultimage-$(CONFIG_SH_7206_SOLUTION_ENGINE)	:= vmlinux
