From: Ben Hutchings <ben@decadent.org.uk>
Date: Wed, 19 Mar 2014 16:08:27 +0000
Subject: can: Avoid ABI change in 3.13.6
Forwarded: not-needed

Hide the new #include's from genksyms.

--- a/include/linux/can/skb.h
+++ b/include/linux/can/skb.h
@@ -11,9 +11,11 @@
 #define CAN_SKB_H
 
 #include <linux/types.h>
-#include <linux/skbuff.h>
 #include <linux/can.h>
+#ifndef __GENKSYMS__
+#include <linux/skbuff.h>
 #include <net/sock.h>
+#endif
 
 /*
  * The struct can_skb_priv is used to transport additional information along
