From: Ben Hutchings <ben@decadent.org.uk>
Date: Sun, 15 Jun 2014 21:41:43 +0100
Subject: drivers/base/platform: Avoid ABI change in 3.14.6
Forwarded: not-needed

Hide the new #include from genksyms.

--- a/drivers/base/platform.c
+++ b/drivers/base/platform.c
@@ -13,7 +13,9 @@
 #include <linux/string.h>
 #include <linux/platform_device.h>
 #include <linux/of_device.h>
+#ifndef __GENKSYMS__
 #include <linux/of_irq.h>
+#endif
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/dma-mapping.h>
