diff --git a/drivers/ata/Kconfig b/drivers/ata/Kconfig
index ae84949..f17c19b 100644
--- a/drivers/ata/Kconfig
+++ b/drivers/ata/Kconfig
@@ -172,7 +172,7 @@ config SATA_SIL
 config SATA_SIS
 	tristate "SiS 964/965/966/180 SATA support"
 	depends on PCI
-	select PATA_SIS
+	select PATA_SIS_STUB
 	help
 	  This option enables support for SiS Serial ATA on
 	  SiS 964/965/966/180 and Parallel ATA on SiS 180.
@@ -618,9 +618,13 @@ config PATA_SIL680
 
 	  If unsure, say N.
 
+config PATA_SIS_STUB
+	tristate
+
 config PATA_SIS
 	tristate "SiS PATA support"
 	depends on PCI
+	select PATA_SIS_STUB
 	help
 	  This option enables support for SiS PATA controllers
 
diff --git a/drivers/ata/Makefile b/drivers/ata/Makefile
index 674965f..c4bce57 100644
--- a/drivers/ata/Makefile
+++ b/drivers/ata/Makefile
@@ -63,7 +63,7 @@ obj-$(CONFIG_PATA_SIL680)	+= pata_sil680.o
 obj-$(CONFIG_PATA_VIA)		+= pata_via.o
 obj-$(CONFIG_PATA_WINBOND)	+= pata_sl82c105.o
 obj-$(CONFIG_PATA_WINBOND_VLB)	+= pata_winbond.o
-obj-$(CONFIG_PATA_SIS)		+= pata_sis.o
+obj-$(CONFIG_PATA_SIS_STUB)	+= pata_sis.o
 obj-$(CONFIG_PATA_TRIFLEX)	+= pata_triflex.o
 obj-$(CONFIG_PATA_IXP4XX_CF)	+= pata_ixp4xx_cf.o
 obj-$(CONFIG_PATA_SCC)		+= pata_scc.o
diff --git a/drivers/ata/pata_sis.c b/drivers/ata/pata_sis.c
index 26345d7..2c3e3ba 100644
--- a/drivers/ata/pata_sis.c
+++ b/drivers/ata/pata_sis.c
@@ -826,13 +826,16 @@ static int sis_init_one (struct pci_dev *pdev, const struct pci_device_id *ent)
 }
 
 static const struct pci_device_id sis_pci_tbl[] = {
+#ifdef CONFIG_PATA_SIS
 	{ PCI_VDEVICE(SI, 0x5513), },	/* SiS 5513 */
 	{ PCI_VDEVICE(SI, 0x5518), },	/* SiS 5518 */
 	{ PCI_VDEVICE(SI, 0x1180), },	/* SiS 1180 */
+#endif
 
 	{ }
 };
 
+#ifdef CONFIG_PATA_SIS
 static struct pci_driver sis_pci_driver = {
 	.name			= DRV_NAME,
 	.id_table		= sis_pci_tbl,
@@ -856,6 +859,7 @@ static void __exit sis_exit(void)
 
 module_init(sis_init);
 module_exit(sis_exit);
+#endif
 
 MODULE_AUTHOR("Alan Cox");
 MODULE_DESCRIPTION("SCSI low-level driver for SiS ATA");
