--- linux-debian-clone.git/.gitignore	2018-02-22 11:37:13.073659326 +0800
+++ deepin-4.14.20/.gitignore	2018-02-22 11:51:06.383793616 +0800
@@ -61,23 +61,11 @@
 /*.spec
 
 #
-# Debian directory (make deb-pkg)
-#
-/debian/
-
-#
 # tar directory (make tar*-pkg)
 #
 /tar-install/
 
 #
-# git files that we don't want to ignore even if they are dot-files
-#
-!.gitignore
-!.mailmap
-!.cocciconfig
-
-#
 # Generated include files
 #
 include/config
@@ -122,3 +110,10 @@
 
 # Kdevelop4
 *.kdev4
+
+#
+# Debian packaging: ignore everything at the top level, since it isn't
+# included in our repository
+#
+/*
+!/debian/
