diff --git a/include/net/inetpeer.h b/include/net/inetpeer.h
index e9ff3fc..7c73e84 100644
--- a/include/net/inetpeer.h
+++ b/include/net/inetpeer.h
@@ -35,7 +35,6 @@ struct inet_peer {
 
 	u32			metrics[RTAX_MAX];
 	u32			rate_tokens;	/* rate limiting for ICMP */
-	int			redirect_genid;
 	unsigned long		rate_last;
 	unsigned long		pmtu_expires;
 	u32			pmtu_orig;
@@ -60,6 +59,10 @@ struct inet_peer {
 	/* following fields might be frequently dirtied */
 	__u32			dtime;	/* the time of last use of not referenced entries */
 	atomic_t		refcnt;
+
+#ifndef __GENKSYMS__
+	int			redirect_genid;
+#endif
 };
 
 void			inet_initpeers(void) __init;
