diff --git a/Makefile b/Makefile
index f3e2065..3cc474a 100644
--- a/Makefile
+++ b/Makefile
@@ -193,25 +193,6 @@ export KBUILD_BUILDHOST := $(SUBARCH)
 ARCH		?= $(SUBARCH)
 CROSS_COMPILE	?=
 
-# Architecture as present in compile.h
-UTS_MACHINE 	:= $(ARCH)
-SRCARCH 	:= $(ARCH)
-
-# Additional ARCH settings for x86
-ifeq ($(ARCH),i386)
-        SRCARCH := x86
-endif
-ifeq ($(ARCH),x86_64)
-        SRCARCH := x86
-endif
-
-# Where to locate arch specific headers
-ifeq ($(ARCH),sparc64)
-       hdr-arch  := sparc
-else
-       hdr-arch  := $(SRCARCH)
-endif
-
 KCONFIG_CONFIG	?= .config
 
 # SHELL used by kbuild
@@ -329,6 +310,27 @@ CFLAGS_KERNEL	=
 AFLAGS_KERNEL	=
 
 
+-include .kernelvariables
+
+# Architecture as present in compile.h
+UTS_MACHINE 	:= $(ARCH)
+SRCARCH 	:= $(ARCH)
+
+# Additional ARCH settings for x86
+ifeq ($(ARCH),i386)
+        SRCARCH := x86
+endif
+ifeq ($(ARCH),x86_64)
+        SRCARCH := x86
+endif
+
+# Where to locate arch specific headers
+ifeq ($(ARCH),sparc64)
+	hdr-arch  := sparc
+else
+	hdr-arch  := $(SRCARCH)
+endif
+
 # Use LINUXINCLUDE when you must reference the include/ directory.
 # Needed to be compatible with the O= option
 LINUXINCLUDE    := -Iinclude \
