diff --git a/Makefile b/Makefile
index edfc2fd..2fb76d7 100644
--- a/Makefile
+++ b/Makefile
@@ -176,9 +176,6 @@ # Note: Some architectures assign CROSS_
 ARCH		?= $(SUBARCH)
 CROSS_COMPILE	?=
 
-# Architecture as present in compile.h
-UTS_MACHINE := $(ARCH)
-
 KCONFIG_CONFIG	?= .config
 
 # SHELL used by kbuild
@@ -298,6 +295,11 @@ LDFLAGS_MODULE  = -r
 CFLAGS_KERNEL	=
 AFLAGS_KERNEL	=
 
+-include .kernelvariables
+
+# Architecture as present in compile.h
+UTS_MACHINE := $(ARCH)
+
 
 # Use LINUXINCLUDE when you must reference the include/ directory.
 # Needed to be compatible with the O= option
