From: Ben Hutchings <ben@decadent.org.uk>
Date: Thu, 31 May 2012 04:04:55 +0100
Subject: [PATCH] mmc: Avoid ABI change in 3.2.19

---
 include/linux/mmc/host.h |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/include/linux/mmc/host.h b/include/linux/mmc/host.h
index deb6282..20caeb0 100644
--- a/include/linux/mmc/host.h
+++ b/include/linux/mmc/host.h
@@ -303,7 +303,6 @@ struct mmc_host {
 
 	unsigned int		sdio_irqs;
 	struct task_struct	*sdio_irq_thread;
-	bool			sdio_irq_pending;
 	atomic_t		sdio_irq_thread_abort;
 
 	mmc_pm_flag_t		pm_flags;	/* requested pm features */
@@ -315,6 +314,10 @@ struct mmc_host {
 #ifdef CONFIG_REGULATOR
 	bool			regulator_enabled; /* regulator state */
 #endif
+#if !(defined(__GENKSYMS__) && defined(CONFIG_REGULATOR))
+	/* Use the hole after regulator_enabled to avoid an ABI change */
+	bool			sdio_irq_pending;
+#endif
 
 	struct dentry		*debugfs_root;
 
-- 
1.7.10

