--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -1869,6 +1869,14 @@
 }
 EXPORT_SYMBOL(netif_device_attach);
 
+/* This function was a bad idea and was removed in 3.2.20 */
+void skb_set_dev(struct sk_buff *skb, struct net_device *dev)
+{
+	WARN_ON_ONCE(!net_eq(dev_net(skb->dev), dev_net(dev)));
+	skb->dev = dev;
+}
+EXPORT_SYMBOL(skb_set_dev);
+
 /*
  * Invalidate hardware checksum when packet is to be mangled, and
  * complete checksum manually on outgoing path.
