From: Ben Hutchings <ben@decadent.org.uk>
Date: Mon, 08 Dec 2014 01:39:31 +0000
Subject: perf: Fix ABI change in 3.16.7-ckt2
Forwarded: not-needed

Hide the new #include from genksyms.


--- a/kernel/events/core.c
+++ b/kernel/events/core.c
@@ -41,7 +41,9 @@
 #include <linux/cgroup.h>
 #include <linux/module.h>
 #include <linux/mman.h>
+#ifndef __GENKSYMS__
 #include <linux/compat.h>
+#endif
 
 #include "internal.h"
 
