diff --git a/arch/powerpc/Makefile b/arch/powerpc/Makefile
index 6238b58..edfbed7 100644
--- a/arch/powerpc/Makefile
+++ b/arch/powerpc/Makefile
@@ -151,13 +151,19 @@
 
 BOOT_TARGETS = zImage zImage.initrd uImage
 
+BOOT_TARGETS_SPECIAL = mkvmlinuz_support_install
+
 PHONY += $(BOOT_TARGETS)
+PHONY += $(BOOT_TARGETS_SPECIAL)
 
 boot := arch/$(ARCH)/boot
 
 $(BOOT_TARGETS): vmlinux
 	$(Q)$(MAKE) ARCH=ppc64 $(build)=$(boot) $(patsubst %,$(boot)/%,$@)
 
+$(BOOT_TARGETS_SPECIAL):
+	$(Q)$(MAKE) $(build)=$(boot) $@
+
 define archhelp
   @echo '* zImage          - Compressed kernel image (arch/$(ARCH)/boot/zImage.*)'
   @echo '  install         - Install kernel using'
diff --git a/arch/powerpc/boot/Makefile b/arch/powerpc/boot/Makefile
index 8378898..dd4c803 100644
--- a/arch/powerpc/boot/Makefile
+++ b/arch/powerpc/boot/Makefile
@@ -235,3 +235,11 @@ clean-kernel := vmlinux.strip vmlinux.bin
 clean-kernel += $(addsuffix .gz,$(clean-kernel))
 # If not absolute clean-files are relative to $(obj).
 clean-files += $(addprefix $(objtree)/, $(clean-kernel))
+
+quiet_cmd_mkvmlinuz = INSTALL mkvmlinuz support files
+	cmd_mkvmlinuz = cp -f $(filter-out FORCE,$^) $(INSTALL_MKVMLINUZ)
+
+mkvmlinuz_support_install: $(wrapperbits)
+	@mkdir -p $(INSTALL_MKVMLINUZ)
+	$(call cmd,mkvmlinuz)
+
