diff --git a/arch/powerpc/Makefile b/arch/powerpc/Makefile
index a00fe72..8486c1c 100644
--- a/arch/powerpc/Makefile
+++ b/arch/powerpc/Makefile
@@ -149,14 +149,19 @@ all: $(KBUILD_IMAGE)
 CPPFLAGS_vmlinux.lds	:= -Upowerpc
 
 BOOT_TARGETS = zImage zImage.initrd uImage
-
+BOOT_SPECIAL_TARGETS = mkvmlinuz_support_install
+  
 PHONY += $(BOOT_TARGETS)
+PHONY += $(BOOT_SPECIAL_TARGETS)
 
 boot := arch/$(ARCH)/boot
 
 $(BOOT_TARGETS): vmlinux
 	$(Q)$(MAKE) ARCH=ppc64 $(build)=$(boot) $(patsubst %,$(boot)/%,$@)
 
+$(BOOT_SPECIAL_TARGETS):
+	$(Q)$(MAKE) ARCH=ppc64 $(build)=$(boot) $@
+
 define archhelp
   @echo '* zImage          - Compressed kernel image (arch/$(ARCH)/boot/zImage.*)'
   @echo '  install         - Install kernel using'
diff --git a/arch/powerpc/boot/Makefile b/arch/powerpc/boot/Makefile
index dc77940..96e6b90 100644
--- a/arch/powerpc/boot/Makefile
+++ b/arch/powerpc/boot/Makefile
@@ -187,3 +187,10 @@ install: $(CONFIGURE) $(image-y)
 clean-files += $(addprefix $(objtree)/, $(obj-boot) vmlinux.strip.gz)
 clean-files += $(addprefix $(objtree)/, $(obj-boot) vmlinux.bin.gz)
 clean-files += $(image-)
+
+quiet_cmd_mkvmlinuz = INSTALL mkvmlinuz support files
+	cmd_mkvmlinuz = cp -f $(filter-out FORCE,$^) $(INSTALL_MKVMLINUZ)
+
+mkvmlinuz_support_install: $(wrapperbits)
+	@mkdir -p $(INSTALL_MKVMLINUZ)
+	$(call cmd,mkvmlinuz)
