From: Ben Hutchings <ben@decadent.org.uk>
Subject: powerpc/machdep: Avoid ABI change in 3.9.6
Forwarded: not-needed

--- a/arch/powerpc/include/asm/machdep.h
+++ b/arch/powerpc/include/asm/machdep.h
@@ -108,8 +108,6 @@ struct machdep_calls {
 	void		(*pcibios_fixup)(void);
 	int		(*pci_probe_mode)(struct pci_bus *);
 	void		(*pci_irq_fixup)(struct pci_dev *dev);
-	int		(*pcibios_root_bridge_prepare)(struct pci_host_bridge
-				*bridge);
 
 	/* To setup PHBs when using automatic OF platform driver for PCI */
 	int		(*pci_setup_phb)(struct pci_controller *host);
@@ -259,6 +257,11 @@ struct machdep_calls {
 	ssize_t (*cpu_probe)(const char *, size_t);
 	ssize_t (*cpu_release)(const char *, size_t);
 #endif
+
+#ifndef __GENKSYMS__
+	int		(*pcibios_root_bridge_prepare)(struct pci_host_bridge
+				*bridge);
+#endif
 };
 
 extern void e500_idle(void);
