From: Ben Hutchings <ben@decadent.org.uk>
Date: Sun, 27 Dec 2015 21:11:09 +0000
Subject: ptrace: Fix ABI change for priv-esc fix
Forwarded: not-needed

Hide the new #include from genksyms.
---
--- a/kernel/ptrace.c
+++ b/kernel/ptrace.c
@@ -20,7 +20,9 @@
 #include <linux/uio.h>
 #include <linux/audit.h>
 #include <linux/pid_namespace.h>
+#ifndef __GENKSYMS__
 #include <linux/user_namespace.h>
+#endif
 #include <linux/syscalls.h>
 #include <linux/uaccess.h>
 #include <linux/regset.h>
