From: Salvatore Bonaccorso <carnil@debian.org>
Date: Fri, 5 Jan 2018 13:36:38 +0100
Subject: Revert "scsi: libsas: allow async aborts"
Bug-Debian: https://bugs.debian.org/882414

---
 drivers/scsi/libsas/sas_scsi_host.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/scsi/libsas/sas_scsi_host.c b/drivers/scsi/libsas/sas_scsi_host.c
index ea8ad06ff582..e81c8590d254 100644
--- a/drivers/scsi/libsas/sas_scsi_host.c
+++ b/drivers/scsi/libsas/sas_scsi_host.c
@@ -491,6 +491,9 @@ int sas_eh_abort_handler(struct scsi_cmnd *cmd)
 	struct Scsi_Host *host = cmd->device->host;
 	struct sas_internal *i = to_sas_internal(host->transportt);
 
+	if (current != host->ehandler)
+		return FAILED;
+
 	if (!i->dft->lldd_abort_task)
 		return FAILED;
 
-- 
2.11.0

