From: Ben Hutchings <ben@decadent.org.uk>
Date: Fri, 11 Jul 2014 09:07:00 +0100
Subject: trace/syscall: Avoid ABI change in 3.14.11

Hide the added #include from genksyms.

--- a/include/trace/syscall.h
+++ b/include/trace/syscall.h
@@ -4,7 +4,9 @@
 #include <linux/tracepoint.h>
 #include <linux/unistd.h>
 #include <linux/ftrace_event.h>
+#ifndef __GENKSYMS__
 #include <linux/thread_info.h>
+#endif
 
 #include <asm/ptrace.h>
 
