From: Ben Hutchings <ben@decadent.org.uk>
Date: Tue, 15 Apr 2014 22:05:04 +0100
Subject: trace: Avoid ABI change in 3.13.8
Forwarded: not-needed

--- a/kernel/trace/trace_events.c
+++ b/kernel/trace/trace_events.c
@@ -27,6 +27,12 @@
 
 DEFINE_MUTEX(event_mutex);
 
+/* Should no longer be used, but kept for compatibility with older modules */
+DEFINE_MUTEX(event_storage_mutex);
+EXPORT_SYMBOL_GPL(event_storage_mutex);
+char event_storage[128];
+EXPORT_SYMBOL_GPL(event_storage);
+
 LIST_HEAD(ftrace_events);
 static LIST_HEAD(ftrace_common_fields);
 
