From: Ben Hutchings <ben@decadent.org.uk>
Date: Mon, 08 Dec 2014 01:34:44 +0000
Subject: truncate: Fix ABI change in 3.16.7-ckt1
Forwarded: not-needed

Hide the new #include from genksyms.

--- a/mm/truncate.c
+++ b/mm/truncate.c
@@ -20,7 +20,9 @@
 #include <linux/buffer_head.h>	/* grr. try_to_release_page,
 				   do_invalidatepage */
 #include <linux/cleancache.h>
+#ifndef __GENKSYMS__
 #include <linux/rmap.h>
+#endif
 #include "internal.h"
 
 static void clear_exceptional_entry(struct address_space *mapping,
