From: Ben Hutchings <ben@decadent.org.uk>
Subject: udp: Fix ABI change in 4.0.6
Forwarded: not-needed

Hide the new #include from genksyms.


--- a/net/ipv4/udp.c
+++ b/net/ipv4/udp.c
@@ -90,7 +90,9 @@
 #include <linux/socket.h>
 #include <linux/sockios.h>
 #include <linux/igmp.h>
+#ifndef __GENKSYMS__
 #include <linux/inetdevice.h>
+#endif
 #include <linux/in.h>
 #include <linux/errno.h>
 #include <linux/timer.h>
