diff --git a/include/linux/usb/hcd.h b/include/linux/usb/hcd.h
index 64cec8d..605092f 100644
--- a/include/linux/usb/hcd.h
+++ b/include/linux/usb/hcd.h
@@ -128,8 +128,10 @@ struct usb_hcd {
 	unsigned		wireless:1;	/* Wireless USB HCD */
 	unsigned		authorized_default:1;
 	unsigned		has_tt:1;	/* Integrated TT in root hub */
+#ifndef __GENKSYMS__
 	unsigned		broken_pci_sleep:1;	/* Don't put the
 			controller in PCI-D3 for system sleep */
+#endif
 
 	int			irq;		/* irq allocated */
 	void __iomem		*regs;		/* device memory/io */
