diff --git a/fs/proc/proc_misc.c b/fs/proc/proc_misc.c
index 92ea774..bbab264 100644
--- a/fs/proc/proc_misc.c
+++ b/fs/proc/proc_misc.c
@@ -254,10 +254,18 @@ static int version_read_proc(char *page, char **start, off_t off,
 {
 	int len;
 
-	/* FIXED STRING! Don't touch! */
 	len = snprintf(page, PAGE_SIZE,
 		"%s version %s"
+#ifdef LINUX_COMPILE_SYSTEM_OFFICIAL
+		" (" LINUX_COMPILE_SYSTEM_DISTRIBUTION " "
+		     LINUX_COMPILE_SYSTEM_VERSION ")"
+		" (" LINUX_COMPILE_SYSTEM_MAINTAINER ")"
+#else
+# ifdef LINUX_COMPILE_SYSTEM_VERSION
+		" (" LINUX_COMPILE_SYSTEM_VERSION ")"
+# endif
 		" (" LINUX_COMPILE_BY "@" LINUX_COMPILE_HOST ")"
+#endif
 		" (" LINUX_COMPILER ")"
 		" %s\n",
 		utsname()->sysname,
diff --git a/init/main.c b/init/main.c
index 2b1cdaa..f860297 100644
--- a/init/main.c
+++ b/init/main.c
@@ -484,7 +484,16 @@ void __init __attribute__((weak)) smp_setup_processor_id(void)
 
 static const char linux_banner[] =
 	"Linux version " UTS_RELEASE
+#ifdef LINUX_COMPILE_SYSTEM_OFFICIAL
+	" (" LINUX_COMPILE_SYSTEM_DISTRIBUTION " "
+	     LINUX_COMPILE_SYSTEM_VERSION ")"
+	" (" LINUX_COMPILE_SYSTEM_MAINTAINER ")"
+#else
+# ifdef LINUX_COMPILE_SYSTEM_VERSION
+	" (" LINUX_COMPILE_SYSTEM_VERSION ")"
+# endif
 	" (" LINUX_COMPILE_BY "@" LINUX_COMPILE_HOST ")"
+#endif
 	" (" LINUX_COMPILER ")"
 	" " UTS_VERSION "\n";
 
diff --git a/init/version.c b/init/version.c
diff --git a/scripts/mkcompile_h b/scripts/mkcompile_h
index d7b8a38..87035a0 100755
--- a/scripts/mkcompile_h
+++ b/scripts/mkcompile_h
@@ -46,16 +46,34 @@ UTS_TRUNCATE="sed -e s/\(.\{1,$UTS_LEN\}\).*/\1/"
 
   echo \#define UTS_VERSION \"`echo $UTS_VERSION | $UTS_TRUNCATE`\"
 
+  if [ -e debian/changelog ]; then
+    DISTRIBUTION=$(lsb_release -is 2>/dev/null)
+    DISTRIBUTION=${DISTRIBUTION:-Debian}
+    MAINTAINER=$(dpkg-parsechangelog | sed -ne 's,^Maintainer: .[^<]*<\([^>]*\)>,\1,p')
+    VERSION=$(dpkg-parsechangelog | awk '/^Version:/ {print $$2}')
+    echo \#define LINUX_COMPILE_SYSTEM_DISTRIBUTION \"$DISTRIBUTION\"
+    echo \#define LINUX_COMPILE_SYSTEM_MAINTAINER \"$MAINTAINER\"
+    echo \#define LINUX_COMPILE_SYSTEM_VERSION \"$VERSION\"
+  fi
+
   echo \#define LINUX_COMPILE_TIME \"`LC_ALL=C LANG=C date +%T`\"
-  echo \#define LINUX_COMPILE_BY \"`whoami`\"
-  echo \#define LINUX_COMPILE_HOST \"`hostname | $UTS_TRUNCATE`\"
 
-  if [ -x /bin/dnsdomainname ]; then
-    echo \#define LINUX_COMPILE_DOMAIN \"`dnsdomainname | $UTS_TRUNCATE`\"
-  elif [ -x /bin/domainname ]; then
-    echo \#define LINUX_COMPILE_DOMAIN \"`domainname | $UTS_TRUNCATE`\"
-  else
+  if [ -e debian/official ]; then
+    echo \#define LINUX_COMPILE_SYSTEM_OFFICIAL
+    echo \#define LINUX_COMPILE_BY \"unknown\"
+    echo \#define LINUX_COMPILE_HOST \"$DISTRIBUTION\"
     echo \#define LINUX_COMPILE_DOMAIN
+  else
+    echo \#define LINUX_COMPILE_BY \"`whoami`\"
+    echo \#define LINUX_COMPILE_HOST \"`hostname | $UTS_TRUNCATE`\"
+
+    if [ -x /bin/dnsdomainname ]; then
+      echo \#define LINUX_COMPILE_DOMAIN \"`dnsdomainname | $UTS_TRUNCATE`\"
+    elif [ -x /bin/domainname ]; then
+      echo \#define LINUX_COMPILE_DOMAIN \"`domainname | $UTS_TRUNCATE`\"
+    else
+      echo \#define LINUX_COMPILE_DOMAIN
+    fi
   fi
 
   echo \#define LINUX_COMPILER \"`$CC -v 2>&1 | tail -n 1`\"
