From: Ben Hutchings <ben@decadent.org.uk>
Date: Tue, 15 Apr 2014 21:56:00 +0100
Subject: vfs: Avoid ABI change in 3.13.9
Forwarded: not-needed

--- a/fs/mount.h
+++ b/fs/mount.h
@@ -18,14 +18,27 @@ struct mnt_pcp {
 	int mnt_writers;
 };
 
+/*
+ * These structures are not exposed to modules, but changes somehow
+ * still affect exported symbol versions!
+ */
+
 struct mountpoint {
+#ifdef __GENKSYMS__
+	struct list_head m_hash; /* not really */
+#else
 	struct hlist_node m_hash;
+#endif
 	struct dentry *m_dentry;
 	int m_count;
 };
 
 struct mount {
+#ifdef __GENKSYMS__
+	struct list_head mnt_hash; /* not really */
+#else
 	struct hlist_node mnt_hash;
+#endif
 	struct mount *mnt_parent;
 	struct dentry *mnt_mountpoint;
 	struct vfsmount mnt;
--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -23,7 +23,9 @@
 #include <linux/uaccess.h>
 #include <linux/proc_ns.h>
 #include <linux/magic.h>
+#ifndef __GENKSYMS__
 #include <linux/bootmem.h>
+#endif
 #include "pnode.h"
 #include "internal.h"
 
