--- linux-2.6.15/init/version.c	2005-03-02 12:39:08 +0100
+++ linux-2.6.15/init/version.c	2006-01-04 19:50:18 +0100
@@ -31,3 +31,18 @@
 	"(" LINUX_COMPILE_BY "@" LINUX_COMPILE_HOST ") "
 #endif
 	"(" LINUX_COMPILER ") " UTS_VERSION "\n";
+
+const char vx_linux_banner[] =
+       "Linux version %s "
+#ifdef LINUX_COMPILE_SYSTEM_OFFICIAL
+       "(" LINUX_COMPILE_SYSTEM_DISTRIBUTION " "
+           LINUX_COMPILE_SYSTEM_VERSION ") "
+       "(" LINUX_COMPILE_SYSTEM_MAINTAINER ") "
+#else
+# ifdef LINUX_COMPILE_SYSTEM_VERSION
+       "(" LINUX_COMPILE_SYSTEM_VERSION ") "
+# endif
+       "(" LINUX_COMPILE_BY "@" LINUX_COMPILE_HOST ") "
+#endif
+       "(" LINUX_COMPILER ") %s\n";
+
