---
 arch/m68k/amiga/platform.c |   18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

--- a/arch/m68k/amiga/platform.c
+++ b/arch/m68k/amiga/platform.c
@@ -12,9 +12,23 @@
 #include <asm/amigahw.h>
 
 
+static struct resource amiga_serial_resources[] = {
+	{
+		/*
+		 *  We request SERDAT and SERPER only, because the serial
+		 *  registers are too spread over the custom register space
+		 */
+		.start	= CUSTOM_PHYSADDR+0x30,
+		.end	= CUSTOM_PHYSADDR+0x33,
+		.flags	= IORESOURCE_MEM,
+	}
+};
+
 static struct platform_device amiga_serial = {
-	.name =	"amiga-serial",
-	.id	= -1,
+	.name		= "amiga-serial",
+	.id		= -1,
+	.num_resources	= ARRAY_SIZE(amiga_serial_resources),
+	.resource	= amiga_serial_resources,
 };
 
 static int __init amiga_init_devices(void)
