---
 arch/m68k/atari/debug.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/arch/m68k/atari/debug.c
+++ b/arch/m68k/atari/debug.c
@@ -20,6 +20,10 @@
 #include <asm/atarihw.h>
 #include <asm/atariints.h>
 
+/* Flag that Modem1 port is already initialized and used */
+int atari_SCC_init_done;
+EXPORT_SYMBOL(atari_SCC_init_done);
+
 /* Can be set somewhere, if a SCC master reset has already be done and should
  * not be repeated; used by kgdb */
 int atari_SCC_reset_done;
@@ -259,6 +263,7 @@ static void __init atari_init_scc_port(i
 	SCC_WRITE(5, reg5 | 8);
 
 	atari_SCC_reset_done = 1;
+	atari_SCC_init_done = 1;
 }
 
 static void __init atari_init_midi_port(int cflag)
