---
 drivers/net/b44.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/net/b44.c
+++ b/drivers/net/b44.c
@@ -2233,7 +2233,9 @@ static void __devexit b44_remove_one(str
 	unregister_netdev(dev);
 	ssb_bus_may_powerdown(sdev->bus);
 	free_netdev(dev);
+#ifdef CONFIG_B44_PCI
 	ssb_pcihost_set_power_state(sdev, PCI_D3hot);
+#endif
 	ssb_set_drvdata(sdev, NULL);
 }
 
@@ -2262,7 +2264,9 @@ static int b44_suspend(struct ssb_device
 		b44_setup_wol(bp);
 	}
 
+#ifdef CONFIG_B44_PCI
 	ssb_pcihost_set_power_state(sdev, PCI_D3hot);
+#endif
 	return 0;
 }
 
