--- a/drivers/staging/speakup/i18n.c	2010-05-18 17:17:16.000000000 +0000
+++ b/drivers/staging/speakup/i18n.c	2010-05-18 17:18:04.000000000 +0000
@@ -1,6 +1,7 @@
 /* Internationalization implementation.  Includes definitions of English
  * string arrays, and the i18n pointer. */
 
+#include <linux/slab.h>
 #include <linux/ctype.h>
 #include <linux/module.h>
 #include <linux/string.h>
--- a/drivers/staging/speakup/selection.c	2010-05-18 17:20:15.000000000 +0000
+++ b/drivers/staging/speakup/selection.c	2010-05-18 17:20:29.000000000 +0000
@@ -2,6 +2,7 @@
 #include <linux/interrupt.h>
 #include <linux/sched.h>
 #include <linux/selection.h>
+#include <linux/slab.h>
 
 #include "speakup.h"
 
--- a/drivers/staging/speakup/kobjects.c	2010-05-18 17:19:00.000000000 +0000
+++ b/drivers/staging/speakup/kobjects.c	2010-05-18 17:19:21.000000000 +0000
@@ -13,6 +13,7 @@
  */
 #include <linux/kernel.h>
 #include <linux/kobject.h>
+#include <linux/slab.h>
 #include <linux/string.h>
 #include <linux/sysfs.h>
 #include <linux/ctype.h>
