From: Ben Hutchings <ben@decadent.org.uk>
Subject: [PATCH] speakup: integrate into kbuild

--- a/drivers/staging/Kconfig
+++ b/drivers/staging/Kconfig
@@ -146,6 +146,8 @@
 source "drivers/staging/mrst-touchscreen/Kconfig"
 
 source "drivers/staging/msm/Kconfig"
+
+source "drivers/staging/speakup/Kconfig"
 
 endif # !STAGING_EXCLUDE_BUILD
 endif # STAGING
--- a/drivers/staging/Makefile
+++ b/drivers/staging/Makefile
@@ -53,3 +53,4 @@
 obj-$(CONFIG_FB_XGI)		+= xgifb/
 obj-$(CONFIG_TOUCHSCREEN_MRSTOUCH)	+= mrst-touchscreen/
 obj-$(CONFIG_MSM_STAGING)	+= msm/
+obj-$(CONFIG_SPEAKUP)		+= speakup/
--- a/drivers/staging/speakup/Kbuild
+++ b/drivers/staging/speakup/Kbuild
@@ -1,4 +1,3 @@
-include $M/allmodule.mk
 obj-$(CONFIG_SPEAKUP_SYNTH_ACNTSA) += speakup_acntsa.o
 obj-$(CONFIG_SPEAKUP_SYNTH_ACNTPC) += speakup_acntpc.o
 obj-$(CONFIG_SPEAKUP_SYNTH_APOLLO) += speakup_apollo.o
