From: Ben Hutchings <ben@decadent.org.uk>
Subject: [PATCH] speakup: integrate into kbuild

--- a/drivers/staging/Kconfig
+++ b/drivers/staging/Kconfig
@@ -126,6 +126,8 @@
 source "drivers/staging/sep/Kconfig"
 
 source "drivers/staging/iio/Kconfig"
+
+source "drivers/staging/speakup/Kconfig"
 
 endif # !STAGING_EXCLUDE_BUILD
 endif # STAGING
--- a/drivers/staging/Makefile
+++ b/drivers/staging/Makefile
@@ -45,3 +45,4 @@
 obj-$(CONFIG_RAR_REGISTER)	+= rar/
 obj-$(CONFIG_DX_SEP)		+= sep/
 obj-$(CONFIG_IIO)		+= iio/
+obj-$(CONFIG_SPEAKUP)		+= speakup/
--- a/drivers/staging/speakup/Kbuild
+++ b/drivers/staging/speakup/Kbuild
@@ -1,4 +1,3 @@
-include $M/allmodule.mk
 obj-$(CONFIG_SPEAKUP_SYNTH_ACNTSA) += speakup_acntsa.o
 obj-$(CONFIG_SPEAKUP_SYNTH_ACNTPC) += speakup_acntpc.o
 obj-$(CONFIG_SPEAKUP_SYNTH_APOLLO) += speakup_apollo.o
