From: Ben Hutchings <ben@decadent.org.uk>
Subject: [PATCH] speakup: Fix references to vc_data::vc_tty

--- a/drivers/staging/speakup/buffers.c
+++ b/drivers/staging/speakup/buffers.c
@@ -30,8 +30,8 @@
 	for (i = 0; i < MAX_NR_CONSOLES; i++) {
 		if (speakup_console[i] && speakup_console[i]->tty_stopped)
 			continue;
-		if ((vc_cons[i].d != NULL) && (vc_cons[i].d->vc_tty != NULL))
-			start_tty(vc_cons[i].d->vc_tty);
+		if ((vc_cons[i].d != NULL) && (vc_cons[i].d->port.tty != NULL))
+			start_tty(vc_cons[i].d->port.tty);
 	}
 	unlock_kernel();
 }
@@ -53,8 +53,8 @@
 		return;
 	}
 	for (i = 0; i < MAX_NR_CONSOLES; i++)
-		if ((vc_cons[i].d != NULL) && (vc_cons[i].d->vc_tty != NULL))
-			stop_tty(vc_cons[i].d->vc_tty);
+		if ((vc_cons[i].d != NULL) && (vc_cons[i].d->port.tty != NULL))
+			stop_tty(vc_cons[i].d->port.tty);
 	if (!in_atomic())
 		unlock_kernel();
 	return;
--- a/drivers/staging/speakup/main.c
+++ b/drivers/staging/speakup/main.c
@@ -2010,7 +2010,7 @@
 		return 0;
 
 	spk_lock(flags);
-	tty = vc->vc_tty;
+	tty = vc->port.tty;
 	if (type >= 0xf0)
 		type -= 0xf0;
 	if (type == KT_PAD &&
