diff --git a/arch/powerpc/Makefile b/arch/powerpc/Makefile
index 01667d1..4d650e1 100644
--- a/arch/powerpc/Makefile
+++ b/arch/powerpc/Makefile
@@ -177,7 +177,7 @@ # Temporary hack until we have migrated 
 include/asm: arch/$(ARCH)/include/asm
 arch/$(ARCH)/include/asm: FORCE
 	$(Q)if [ ! -d arch/$(ARCH)/include ]; then mkdir -p arch/$(ARCH)/include; fi
-	$(Q)ln -fsn $(srctree)/include/asm-$(OLDARCH) arch/$(ARCH)/include/asm
+	$(Q)ln -fsn $(if $(KBUILD_SRC),$(KBUILD_SRC),../../..)/include/asm-$(OLDARCH) arch/$(ARCH)/include/asm
 endif
 
 # Use the file '.tmp_gas_check' for binutils tests, as gas won't output
diff --git a/arch/ppc/Makefile b/arch/ppc/Makefile
index 0db66dc..8523a15 100644
--- a/arch/ppc/Makefile
+++ b/arch/ppc/Makefile
@@ -116,7 +116,7 @@ # Temporary hack until we have migrated 
 include/asm: arch/$(ARCH)/include/asm
 arch/$(ARCH)/include/asm:
 	$(Q)if [ ! -d arch/$(ARCH)/include ]; then mkdir -p arch/$(ARCH)/include; fi
-	$(Q)ln -fsn $(srctree)/include/asm-powerpc arch/$(ARCH)/include/asm
+	$(Q)ln -fsn $(if $(KBUILD_SRC),$(KBUILD_SRC),../../..)/include/asm-powerpc arch/$(ARCH)/include/asm
 
 # Use the file '.tmp_gas_check' for binutils tests, as gas won't output
 # to stdout and these checks are run even on install targets.
