From: Steve French <smfrench@gmail.com>
Date: Wed, 19 Oct 2011 21:27:11 -0500
Subject: [PATCH] [CIFS] Fixup trivial checkpatch warning

commit b957ae9c53d5715a07f8bac644d8ff0a407c7e07 upstream.

Signed-off-by: Steve French <smfrench@gmail.com>
---
 include/linux/freezer.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/include/linux/freezer.h b/include/linux/freezer.h
index 3672f73..a155561 100644
--- a/include/linux/freezer.h
+++ b/include/linux/freezer.h
@@ -143,7 +143,7 @@ static inline void set_freezable_with_signal(void)
 ({									\
 	int __retval;							\
 	do {								\
-		__retval = wait_event_killable(wq, 			\
+		__retval = wait_event_killable(wq,			\
 				(condition) || freezing(current));	\
 		if (__retval && !freezing(current))			\
 			break;						\
-- 
1.7.7.2

