From: Ben Hutchings <ben@decadent.org.uk>
Date: Tue, 07 Jul 2015 17:23:54 +0100
Subject: DocBook: generate consistent IDs
Bug-Debian: https://bugs.debian.org/769844

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
---
--- a/Documentation/DocBook/Makefile
+++ b/Documentation/DocBook/Makefile
@@ -69,6 +69,7 @@ DOCPROC   = $(objtree)/scripts/docproc
 
 XMLTOFLAGS = -m $(srctree)/$(src)/stylesheet.xsl
 XMLTOFLAGS += --skip-validation
+XMLTOFLAGS += --stringparam generate.consistent.ids=1
 
 ###
 # DOCPROC is used for two purposes:
