From: Florian Fainelli <ffainelli@freebox.fr>
Subject: [PATCH] PHY: fix typo in bcm63xx PHY driver table
Date: Fri, 9 Apr 2010 13:04:45 +0200

commit 0de8655ab9181cbaca82aa60402b14118e06d030 upstream.

Signed-off-by: Florian Fainelli <ffainelli@freebox.fr>
---
diff --git a/drivers/net/phy/bcm63xx.c b/drivers/net/phy/bcm63xx.c
index ac5e498..c128156 100644
--- a/drivers/net/phy/bcm63xx.c
+++ b/drivers/net/phy/bcm63xx.c
@@ -137,4 +137,4 @@ static struct mdio_device_id bcm63xx_tbl[] = {
 	{ }
 };
 
-MODULE_DEVICE_TABLE(mdio, bcm64xx_tbl);
+MODULE_DEVICE_TABLE(mdio, bcm63xx_tbl);
