From: Ben Hutchings <ben@decadent.org.uk>
Date: Mon, 06 Jul 2015 23:34:54 +0100
Subject: remove timestamps from compressed man pages
Bug-Debian: https://bugs.debian.org/769844

--- a/Documentation/DocBook/Makefile
+++ b/Documentation/DocBook/Makefile
@@ -56,7 +56,7 @@ htmldocs: $(HTML)
 
 MAN := $(patsubst %.xml, %.9, $(BOOKS))
 mandocs: $(MAN)
-	find $(obj)/man -name '*.9' | xargs gzip -f
+	find $(obj)/man -name '*.9' | xargs gzip -nf
 
 installmandocs: mandocs
 	mkdir -p /usr/local/man/man9/
