From: Ben Hutchings <ben@decadent.org.uk>
Date: Fri, 30 Oct 2015 10:22:55 +0000
Subject: selftests: breakpoint: Actually build it

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
---
--- a/tools/testing/selftests/breakpoints/Makefile
+++ b/tools/testing/selftests/breakpoints/Makefile
@@ -6,7 +6,7 @@ ifeq ($(ARCH),x86)
 TEST_PROGS := breakpoint_test
 endif
 
-all:
+all: $(TEST_PROGS)
 
 include ../lib.mk
 
