From: Matthew Garrett <mjg@redhat.com>
Date: Wed Apr 21 14:13:22 2010 UTC
Subject: Add EC path for Thinkpad X100


diff -up linux-2.6.32.noarch/drivers/platform/x86/thinkpad_acpi.c.mjg linux-2.6.32.noarch/drivers/platform/x86/thinkpad_acpi.c
--- linux-2.6.32.noarch/drivers/platform/x86/thinkpad_acpi.c.mjg	2010-04-21 10:02:53.658034129 -0400
+++ linux-2.6.32.noarch/drivers/platform/x86/thinkpad_acpi.c	2010-04-21 10:03:30.402030108 -0400
@@ -491,6 +491,7 @@ TPACPI_HANDLE(ec, root, "\\_SB.PCI0.ISA.
 	   "\\_SB.PCI0.ISA.EC",	/* A21e, A2xm/p, T20-22, X20-21 */
 	   "\\_SB.PCI0.AD4S.EC0",	/* i1400, R30 */
 	   "\\_SB.PCI0.ICH3.EC0",	/* R31 */
+	   "\\_SB.PCI0.LPC0.EC",	/* X100e */
 	   "\\_SB.PCI0.LPC.EC",	/* all others */
 	   );
 
