diff --git a/net/wireless/reg.c b/net/wireless/reg.c
index 855bff4..5969480 100644
--- a/net/wireless/reg.c
+++ b/net/wireless/reg.c
@@ -29,7 +29,7 @@
 #include <net/wireless.h>
 #include "core.h"
 
-static char *ieee80211_regdom = "US";
+static char *ieee80211_regdom = "EU";
 module_param(ieee80211_regdom, charp, 0444);
 MODULE_PARM_DESC(ieee80211_regdom, "IEEE 802.11 regulatory domain code");
 
