diff -r d24540ecc512 arch/i386/kernel/swiotlb.c
--- a/arch/i386/kernel/swiotlb.c	Tue Jan 23 23:37:22 2007 +0100
+++ b/arch/i386/kernel/swiotlb.c	Thu Mar 22 10:38:34 2007 +0100
@@ -227,9 +227,12 @@ __sync_single(struct phys_addr buffer, c
 		char *dev, *host, *kmp;
 		len = size;
 		while (len != 0) {
+			unsigned long flags;
+
 			if (((bytes = len) + buffer.offset) > PAGE_SIZE)
 				bytes = PAGE_SIZE - buffer.offset;
-			kmp  = kmap_atomic(buffer.page, KM_SWIOTLB);
+			local_irq_save(flags); /* protects KM_BOUNCE_READ */
+			kmp  = kmap_atomic(buffer.page, KM_BOUNCE_READ);
 			dev  = dma_addr + size - len;
 			host = kmp + buffer.offset;
 			if (dir == DMA_FROM_DEVICE) {
@@ -237,7 +240,8 @@ __sync_single(struct phys_addr buffer, c
 					/* inaccessible */;
 			} else
 				memcpy(dev, host, bytes);
-			kunmap_atomic(kmp, KM_SWIOTLB);
+			kunmap_atomic(kmp, KM_BOUNCE_READ);
+			local_irq_restore(flags);
 			len -= bytes;
 			buffer.page++;
 			buffer.offset = 0;
