From 8d7bff2d72660d9d60aa371ae3d1356bbf329a09 Mon Sep 17 00:00:00 2001
From: Chris Wright <chrisw@sous-sol.org>
Date: Thu, 2 Apr 2009 13:55:27 -0700
Subject: [PATCH 90/90] Linux 2.6.29.1

---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 1ab3ebf..cdb1133 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 29
-EXTRAVERSION =
+EXTRAVERSION = .1
 NAME = Temporary Tasmanian Devil
 
 # *DOCUMENTATION*
-- 
1.6.2.3

