From: Ben Hutchings <ben@decadent.org.uk>
Subject: [PATCH] aufs: mark as staging

I really don't want to support this.

--- a/fs/aufs/module.c
+++ b/fs/aufs/module.c
@@ -89,6 +89,7 @@
 MODULE_DESCRIPTION(AUFS_NAME
 	" -- Advanced multi layered unification filesystem");
 MODULE_VERSION(AUFS_VERSION);
+MODULE_INFO(staging, "Y");
 
 /* this module parameter has no meaning when SYSFS is disabled */
 int sysaufs_brs = 1;
