From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Tue, 24 Jun 2014 15:03:18 +0200
Subject: [126/249] staging: rtl8723au: Remove unused IE offset #defines
Origin: https://git.kernel.org/linus/5b5cf9941fd01da487e3324668d2237226fdeb23

Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 drivers/staging/rtl8723au/include/wifi.h | 8 --------
 1 file changed, 8 deletions(-)

diff --git a/drivers/staging/rtl8723au/include/wifi.h b/drivers/staging/rtl8723au/include/wifi.h
index bd581ec..d07fd12 100644
--- a/drivers/staging/rtl8723au/include/wifi.h
+++ b/drivers/staging/rtl8723au/include/wifi.h
@@ -23,14 +23,6 @@
  */
 #define	WiFiNavUpperUs		30000	/*  30 ms */
 
-#define _ASOCREQ_IE_OFFSET_		4	/*  excluding wlan_hdr */
-#define	_ASOCRSP_IE_OFFSET_		6
-#define _REASOCREQ_IE_OFFSET_		10
-#define _REASOCRSP_IE_OFFSET_		6
-#define _PROBEREQ_IE_OFFSET_		0
-#define	_PROBERSP_IE_OFFSET_		12
-#define _AUTH_IE_OFFSET_		6
-#define _DEAUTH_IE_OFFSET_		0
 #define _BEACON_IE_OFFSET_		12
 
 #define _FIXED_IE_LENGTH_		_BEACON_IE_OFFSET_
