From e5b457c03b8163cf4c791521a3703a627090ff8c Mon Sep 17 00:00:00 2001
From: Bartlomiej Zolnierkiewicz <bzolnier@gmail.com>
Date: Wed, 21 Oct 2009 22:44:05 +0200
Subject: [PATCH 07/42] Staging: rt28x0: remove private RTPRIV_IOCTL_SET ioctl

commit 468d0c507b9c11a2e3ae95ff149eb551d05be8b9 upstream.

Signed-off-by: Bartlomiej Zolnierkiewicz <bzolnier@gmail.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---
 drivers/staging/rt2860/sta_ioctl.c |    6 +-----
 1 files changed, 1 insertions(+), 5 deletions(-)

diff --git a/drivers/staging/rt2860/sta_ioctl.c b/drivers/staging/rt2860/sta_ioctl.c
index e5e2942..b6ba561 100644
--- a/drivers/staging/rt2860/sta_ioctl.c
+++ b/drivers/staging/rt2860/sta_ioctl.c
@@ -2978,6 +2978,7 @@ INT rt28xx_sta_ioctl(
 		case SIOCSIWRETRY:	//set retry limits and lifetime
 		case RT_PRIV_IOCTL:
 		case RT_PRIV_IOCTL_EXT:
+		case RTPRIV_IOCTL_SET:
 			Status = -EOPNOTSUPP;
 			break;
 		case SIOCGIWPRIV:
@@ -2990,11 +2991,6 @@ INT rt28xx_sta_ioctl(
 					Status = -EFAULT;
 			}
 			break;
-		case RTPRIV_IOCTL_SET:
-			if(access_ok(VERIFY_READ, wrq->u.data.pointer, wrq->u.data.length) != TRUE)
-				break;
-			rt_ioctl_setparam(net_dev, NULL, NULL, wrq->u.data.pointer);
-			break;
 		case RTPRIV_IOCTL_GSITESURVEY:
 			RTMPIoctlGetSiteSurvey(pAd, wrq);
 		    break;
-- 
1.7.1

