From: Michael Grzeschik <m.grzeschik@pengutronix.de>
Date: Thu, 11 Apr 2013 10:13:17 +0000
Subject: ARM: dts: imx: imx53-qsb.dts: enable usbotg and usbh1
Origin: https://git.linaro.org/gitweb?p=people/shawnguo/linux-2.6.git;a=commit;h=c9cb1ec6e2a15e744a6481cb14730a5812471e71

Signed-off-by: Michael Grzeschik <m.grzeschik@pengutronix.de>
Signed-off-by: Shawn Guo <shawn.guo@linaro.org>
---
(limited to 'arch/arm/boot/dts')

diff --git a/arch/arm/boot/dts/imx53-qsb.dts b/arch/arm/boot/dts/imx53-qsb.dts
index 8f0e9ae..160d1bc 100644
--- a/arch/arm/boot/dts/imx53-qsb.dts
+++ b/arch/arm/boot/dts/imx53-qsb.dts
@@ -268,3 +268,11 @@
 	phy-reset-gpios = <&gpio7 6 0>;
 	status = "okay";
 };
+
+&usbh1 {
+       status = "okay";
+};
+
+&usbotg {
+       status = "okay";
+};
--
cgit v0.9.1
