Index: linux-2.6.22-rc3-armeb/arch/arm/mach-ixp4xx/common.c
===================================================================
--- linux-2.6.22-rc3-armeb.orig/arch/arm/mach-ixp4xx/common.c	2007-05-27 07:07:26.000000000 -0700
+++ linux-2.6.22-rc3-armeb/arch/arm/mach-ixp4xx/common.c	2007-05-27 07:08:24.000000000 -0700
@@ -486,3 +486,17 @@
 	clockevents_register_device(&clockevent_ixp4xx);
 	return 0;
 }
+
+/* fuses */
+
+u32 ixp4xx_read_fuses(void)
+{
+	unsigned int fuses = ~*IXP4XX_EXP_CFG2;
+	fuses &= ~IXP4XX_FUSE_RESERVED;
+	if (!cpu_is_ixp46x())
+		fuses &= ~IXP4XX_FUSE_IXP46X_ONLY;
+
+	return fuses;
+}
+
+EXPORT_SYMBOL(ixp4xx_read_fuses);
Index: linux-2.6.22-rc3-armeb/include/asm-arm/arch-ixp4xx/platform.h
===================================================================
--- linux-2.6.22-rc3-armeb.orig/include/asm-arm/arch-ixp4xx/platform.h	2007-05-27 07:07:26.000000000 -0700
+++ linux-2.6.22-rc3-armeb/include/asm-arm/arch-ixp4xx/platform.h	2007-05-27 10:18:14.000000000 -0700
@@ -120,6 +120,47 @@
 extern int ixp4xx_setup(int nr, struct pci_sys_data *sys);
 extern struct pci_bus *ixp4xx_scan_bus(int nr, struct pci_sys_data *sys);
 
+/* Fuse definitions and functions */
+
+/* Fuse Bits of IXP_EXP_CFG2 */
+#define IXP4XX_FUSE_RCOMP               (1 << 0)
+#define IXP4XX_FUSE_USB_DEVICE          (1 << 1)
+#define IXP4XX_FUSE_HASH                (1 << 2)
+#define IXP4XX_FUSE_AES                 (1 << 3)
+#define IXP4XX_FUSE_DES                 (1 << 4)
+#define IXP4XX_FUSE_HDLC                (1 << 5)
+#define IXP4XX_FUSE_AAL                 (1 << 6)
+#define IXP4XX_FUSE_HSS                 (1 << 7)
+#define IXP4XX_FUSE_UTOPIA              (1 << 8)
+#define IXP4XX_FUSE_NPEB_ETH0           (1 << 9)
+#define IXP4XX_FUSE_NPEC_ETH            (1 << 10)
+#define IXP4XX_FUSE_RESET_NPEA          (1 << 11)
+#define IXP4XX_FUSE_RESET_NPEB          (1 << 12)
+#define IXP4XX_FUSE_RESET_NPEC          (1 << 13)
+#define IXP4XX_FUSE_PCI                 (1 << 14)
+#define IXP4XX_FUSE_ECC_TIMESYNC        (1 << 15)
+#define IXP4XX_FUSE_UTOPIA_PHY_LIMIT    (3 << 16)
+#define IXP4XX_FUSE_USB_HOST            (1 << 18)
+#define IXP4XX_FUSE_NPEA_ETH            (1 << 19)
+#define IXP4XX_FUSE_NPEB_ETH_1_TO_3     (1 << 20)
+#define IXP4XX_FUSE_RSA                 (1 << 21)
+#define IXP4XX_FUSE_XSCALE_MAX_FREQ     (3 << 22)
+#define IXP4XX_FUSE_RESERVED            (0xFF << 24)
+
+#define IXP4XX_FUSE_IXP46X_ONLY (IXP4XX_FUSE_ECC_TIMESYNC |             \
+                                 IXP4XX_FUSE_USB_HOST |                 \
+                                 IXP4XX_FUSE_NPEA_ETH |                 \
+                                 IXP4XX_FUSE_NPEB_ETH_1_TO_3 |          \
+                                 IXP4XX_FUSE_RSA |                      \
+                                 IXP4XX_FUSE_XSCALE_MAX_FREQ)
+
+extern u32 ixp4xx_read_fuses(void);
+
+static inline void ixp4xx_write_fuses(u32 value)
+{
+        *IXP4XX_EXP_CFG2 = ~value;
+}
+
 /*
  * GPIO-functions
  */
Index: linux-2.6.22-rc3-armeb/arch/arm/kernel/setup.c
===================================================================
--- linux-2.6.22-rc3-armeb.orig/arch/arm/kernel/setup.c	2007-05-27 07:07:26.000000000 -0700
+++ linux-2.6.22-rc3-armeb/arch/arm/kernel/setup.c	2007-05-27 07:08:24.000000000 -0700
@@ -60,6 +60,8 @@
 extern void _stext, _text, _etext, __data_start, _edata, _end;
 
 unsigned int processor_id;
+EXPORT_SYMBOL(processor_id);
+
 unsigned int __machine_arch_type;
 EXPORT_SYMBOL(__machine_arch_type);
 
