diff -Naurp linux-2.6.22-rc4.orig/arch/arm/mach-ixp4xx/nslu2-setup.c linux-2.6.22-rc4/arch/arm/mach-ixp4xx/nslu2-setup.c
--- linux-2.6.22-rc4.orig/arch/arm/mach-ixp4xx/nslu2-setup.c	2007-06-05 20:37:07.000000000 -0600
+++ linux-2.6.22-rc4/arch/arm/mach-ixp4xx/nslu2-setup.c	2007-06-05 21:12:36.000000000 -0600
@@ -17,6 +17,7 @@
 #include <linux/kernel.h>
 #include <linux/serial.h>
 #include <linux/serial_8250.h>
+#include <linux/mtd/mtd.h>
 #include <linux/leds.h>
 
 #include <asm/mach-types.h>
@@ -171,8 +172,44 @@ static struct sys_timer nslu2_timer = {
     .init   = nslu2_timer_init,
 };
 
+/*
+ * When the RedBoot partition is added the MAC address is read from
+ * it.
+ */
+static void nslu2_flash_add(struct mtd_info *mtd) {
+	if (strcmp(mtd->name, "RedBoot") == 0) {
+		size_t retlen;
+		u_char mac[6];
+
+		/* The MAC is at a known offset... */
+		if (mtd->read(mtd, 0x3FFB0, 6, &retlen, mac) == 0 && retlen == 6) {
+			printk(KERN_INFO "NSLU2 MAC: %.2x:%.2x:%.2x:%.2x:%.2x:%.2x\n",
+				mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+			memcpy(&plat_mac0.hwaddr, mac, 6);
+		} else {
+			printk(KERN_ERR "NSLU2 MAC: read failed\n");
+		}
+	}
+}
+
+/*
+ * Nothing to do on remove at present.
+ */
+static void nslu2_flash_remove(struct mtd_info *mtd) {
+}
+
+static struct mtd_notifier nslu2_flash_notifier = {
+	.add = nslu2_flash_add,
+	.remove = nslu2_flash_remove,
+};
+
 static void __init nslu2_init(void)
 {
+	/* The flash has an ethernet MAC embedded in it which we need,
+	 * that is all this notifier does.
+	 */
+	register_mtd_user(&nslu2_flash_notifier);
+
 	ixp4xx_sys_init();
 
 	nslu2_flash_resource.start = IXP4XX_EXP_BUS_BASE(0);
