Index: linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/nslu2-setup.c
===================================================================
--- linux-2.6.22-rc1-armeb.orig/arch/arm/mach-ixp4xx/nslu2-setup.c	2007-05-16 08:10:47.000000000 -0700
+++ linux-2.6.22-rc1-armeb/arch/arm/mach-ixp4xx/nslu2-setup.c	2007-05-16 08:15:14.000000000 -0700
@@ -140,6 +140,22 @@
 	.resource		= nslu2_uart_resources,
 };
 
+/* Built-in 10/100 Ethernet MAC interfaces */
+static struct mac_plat_info nslu2_plat_mac[] = {
+	{
+		.phy		= 1,
+		.rxq		= 3,
+	}
+};
+
+static struct platform_device nslu2_mac[] = {
+	{
+		.name			= "ixp4xx_eth",
+		.id			= IXP4XX_ETH_NPEB,
+		.dev.platform_data	= nslu2_plat_mac,
+	}
+};
+
 static struct platform_device *nslu2_devices[] __initdata = {
 	&nslu2_i2c_gpio,
 	&nslu2_flash,
@@ -147,6 +163,7 @@
 #ifdef CONFIG_LEDS_IXP4XX
 	&nslu2_leds,
 #endif
+	&nslu2_mac[0],
 };
 
 static void nslu2_power_off(void)
