Index: linux-2.6.18/arch/arm/mach-ixp4xx/nslu2-setup.c
===================================================================
--- linux-2.6.18.orig/arch/arm/mach-ixp4xx/nslu2-setup.c	2006-10-01 23:12:35.000000000 +0200
+++ linux-2.6.18/arch/arm/mach-ixp4xx/nslu2-setup.c	2006-10-01 23:12:40.000000000 +0200
@@ -139,6 +139,28 @@
 	.resource		= nslu2_uart_resources,
 };
 
+static struct resource res_mac0 = {
+	.start		= IXP4XX_EthB_BASE_PHYS,
+	.end		= IXP4XX_EthB_BASE_PHYS + 0x1ff,
+	.flags          = IORESOURCE_MEM,
+};
+
+static struct mac_plat_info plat_mac0 = {
+	.npe_id		= 1,
+	.phy_id		= 1,
+	.eth_id		= 0,
+	.rxq_id		= 27,
+	.txq_id		= 24,
+};
+
+static struct platform_device mac0 = {
+	.name           = "ixp4xx_mac",
+	.id             = 0,
+	.dev.platform_data = &plat_mac0,
+	.num_resources  = 1,
+	.resource       = &res_mac0,
+};
+
 static struct platform_device *nslu2_devices[] __initdata = {
 	&nslu2_i2c_controller,
 	&nslu2_flash,
@@ -146,6 +168,7 @@
 #ifdef CONFIG_LEDS_IXP4XX
 	&nslu2_leds,
 #endif
+	&mac0
 };
 
 static void nslu2_power_off(void)
