--- linux-source-2.6.26/drivers/misc/hpilo.c~	2008-08-06 13:22:32.000000000 -0600
+++ linux-source-2.6.26/drivers/misc/hpilo.c	2008-08-06 13:22:42.000000000 -0600
@@ -688,7 +688,7 @@
 	for (minor = 0 ; minor < MAX_CCB; minor++) {
 		struct device *dev;
 		dev = device_create(ilo_class, &pdev->dev,
-				    MKDEV(ilo_major, minor), NULL,
+				    MKDEV(ilo_major, minor),
 				    "hpilo!d%dccb%d", devnum, minor);
 		if (IS_ERR(dev))
 			dev_err(&pdev->dev, "Could not create files\n");
