diff -ur source-s390-vserver.orig/fs/namespace.c source-s390-vserver/fs/namespace.c
--- source-s390-vserver.orig/fs/namespace.c	2006-08-12 16:51:51.000000000 +0200
+++ source-s390-vserver/fs/namespace.c	2006-08-12 16:53:26.000000000 +0200
@@ -986,6 +986,9 @@
 	if (!mnt)
 		goto out;
 
+	if (!capable(CAP_SYS_ADMIN) && (old_nd.mnt->mnt_flags & MNT_NODEV))
+		mnt_flags |= MNT_NODEV;
+
 	mnt->mnt_flags = mnt_flags;
 	if (flags & MS_XID) {
 		mnt->mnt_xid = xid;
@@ -1028,6 +1031,9 @@
 	if (nd->dentry != nd->mnt->mnt_root)
 		return -EINVAL;
 
+	if (!capable(CAP_SYS_ADMIN))
+		mnt_flags |= MNT_NODEV;
+
 	down_write(&sb->s_umount);
 	err = do_remount_sb(sb, flags, data, 0);
 	if (!err)
@@ -1136,6 +1142,9 @@
 	if (!vx_capable(CAP_SYS_ADMIN, VXC_SECURE_MOUNT))
 		return -EPERM;
 
+	if (!capable(CAP_SYS_ADMIN))
+		mnt_flags |= MNT_NODEV;
+
 	mnt = do_kern_mount(type, flags, name, data);
 	if (IS_ERR(mnt))
 		return PTR_ERR(mnt);
@@ -1409,8 +1418,6 @@
 	if (flags & MS_NODIRATIME)
 		mnt_flags |= MNT_NODIRATIME;
 
-	if (!capable(CAP_SYS_ADMIN))
-		mnt_flags |= MNT_NODEV;
 	flags &= ~(MS_NOSUID | MS_NOEXEC | MS_NODEV | MS_ACTIVE |
 		   MS_NOATIME | MS_NODIRATIME);
 
