diff -urpN a/arch/s390/include/asm/unistd.h b/arch/s390/include/asm/unistd.h
--- a/arch/s390/include/asm/unistd.h	2009-12-03 03:51:21.000000000 +0000
+++ b/arch/s390/include/asm/unistd.h	2009-12-29 18:41:40.000000000 +0000
@@ -202,7 +202,7 @@
 #define __NR_clock_gettime	(__NR_timer_create+6)
 #define __NR_clock_getres	(__NR_timer_create+7)
 #define __NR_clock_nanosleep	(__NR_timer_create+8)
-/* Number 263 is reserved for vserver */
+#define __NR_vserver		263
 #define __NR_statfs64		265
 #define __NR_fstatfs64		266
 #define __NR_remap_file_pages	267
diff -urpN a/fs/exec.c b/fs/exec.c
--- a/fs/exec.c	2009-12-29 17:57:12.000000000 +0000
+++ b/fs/exec.c	2009-12-29 18:46:32.000000000 +0000
@@ -56,6 +56,7 @@
 #include <linux/fsnotify.h>
 #include <linux/fs_struct.h>
 #include <linux/pipe_fs_i.h>
+#include <linux/vs_memory.h>
 
 #include <asm/uaccess.h>
 #include <asm/mmu_context.h>
diff -urpN a/mm/memory.c b/mm/memory.c
--- a/mm/memory.c	2009-12-29 17:57:12.000000000 +0000
+++ b/mm/memory.c	2009-12-29 18:44:58.000000000 +0000
@@ -56,7 +56,7 @@
 #include <linux/kallsyms.h>
 #include <linux/swapops.h>
 #include <linux/elf.h>
-// #include <linux/vs_memory.h>
+#include <linux/vs_memory.h>
 
 #include <asm/io.h>
 #include <asm/pgalloc.h>
diff -urpN a/mm/mmap.c b/mm/mmap.c
--- a/mm/mmap.c	2009-12-29 17:57:12.000000000 +0000
+++ b/mm/mmap.c	2009-12-29 18:45:42.000000000 +0000
@@ -29,6 +29,7 @@
 #include <linux/rmap.h>
 #include <linux/mmu_notifier.h>
 #include <linux/perf_event.h>
+#include <linux/vs_memory.h>
 
 #include <asm/uaccess.h>
 #include <asm/cacheflush.h>
