diff -urpN a/arch/s390/include/asm/tlb.h b/arch/s390/include/asm/tlb.h
--- a/arch/s390/include/asm/tlb.h	2009-12-02 20:51:21.000000000 -0700
+++ b/arch/s390/include/asm/tlb.h	2009-12-30 01:46:05.000000000 -0700
@@ -23,6 +23,7 @@
 
 #include <linux/mm.h>
 #include <linux/swap.h>
+#include <linux/vs_memory.h>
 #include <asm/processor.h>
 #include <asm/pgalloc.h>
 #include <asm/smp.h>
diff -urpN a/arch/s390/include/asm/unistd.h b/arch/s390/include/asm/unistd.h
--- a/arch/s390/include/asm/unistd.h	2009-12-02 20:51:21.000000000 -0700
+++ b/arch/s390/include/asm/unistd.h	2009-12-30 01:42:52.000000000 -0700
@@ -202,7 +202,7 @@
 #define __NR_clock_gettime	(__NR_timer_create+6)
 #define __NR_clock_getres	(__NR_timer_create+7)
 #define __NR_clock_nanosleep	(__NR_timer_create+8)
-/* Number 263 is reserved for vserver */
+#define __NR_vserver		263
 #define __NR_statfs64		265
 #define __NR_fstatfs64		266
 #define __NR_remap_file_pages	267
