diff --git a/include/asm-s390/tlb.h b/include/asm-s390/tlb.h
index 3d8a96d..656a647 100644
--- a/include/asm-s390/tlb.h
+++ b/include/asm-s390/tlb.h
@@ -27,6 +27,7 @@
 #include <asm/pgalloc.h>
 #include <asm/smp.h>
 #include <asm/tlbflush.h>
+#include <linux/vs_memory.h>
 
 #ifndef CONFIG_SMP
 #define TLB_NR_PTRS	1
diff --git a/include/linux/vserver/debug.h b/include/linux/vserver/debug.h
index 2c989b5..c43bbb9 100644
--- a/include/linux/vserver/debug.h
+++ b/include/linux/vserver/debug.h
@@ -11,9 +11,6 @@
 #define VXF_DEV		"%p[%lu,%d:%d]"
 
 
-#define __FUNC__	__func__
-
-
 #define vxd_path(p)						\
 	({ static char _buffer[PATH_MAX];			\
 	   d_path(p, _buffer, sizeof(_buffer)); })
diff --git a/include/linux/vserver/pid.h b/include/linux/vserver/pid.h
index eb0b9cb..c76a6ee 100644
--- a/include/linux/vserver/pid.h
+++ b/include/linux/vserver/pid.h
@@ -4,7 +4,7 @@
 /* pid faking stuff */
 
 #define vx_info_map_pid(v, p) \
-	__vx_info_map_pid((v), (p), __FUNC__, __FILE__, __LINE__)
+	__vx_info_map_pid((v), (p), __func__, __FILE__, __LINE__)
 #define vx_info_map_tgid(v,p)  vx_info_map_pid(v,p)
 #define vx_map_pid(p) vx_info_map_pid(current->vx_info, p)
 #define vx_map_tgid(p) vx_map_pid(p)
@@ -27,7 +27,7 @@ static inline int __vx_info_map_pid(struct vx_info *vxi, int pid,
 }
 
 #define vx_info_rmap_pid(v, p) \
-	__vx_info_rmap_pid((v), (p), __FUNC__, __FILE__, __LINE__)
+	__vx_info_rmap_pid((v), (p), __func__, __FILE__, __LINE__)
 #define vx_rmap_pid(p) vx_info_rmap_pid(current->vx_info, p)
 #define vx_rmap_tgid(p) vx_rmap_pid(p)
 
