From: Thorsten Glaser <tg@debian.org>
Description: m68k: handle Atari interrupts in multi-platform kernels
Origin: http://thread.gmane.org/gmane.linux.debian.devel.kernel/91476/focus=5003
Forwarded: http://thread.gmane.org/gmane.linux.ports.m68k/5009

--- a/arch/m68k/include/asm/irqflags.h
+++ b/arch/m68k/include/asm/irqflags.h
@@ -67,6 +67,10 @@ static inline void arch_local_irq_restor
 
 static inline bool arch_irqs_disabled_flags(unsigned long flags)
 {
+	if (MACH_IS_ATARI) {
+		/* Ignore HSYNC = ipl 2 on Atari */
+		return (flags & ~(ALLOWINT | 0x200)) != 0;
+	}
 	return (flags & ~ALLOWINT) != 0;
 }
 
