Index: linux-2.6-2.6.21/drivers/parport/Kconfig
===================================================================
--- linux-2.6-2.6.21.orig/drivers/parport/Kconfig	2007-05-13 20:53:39.000000000 +0200
+++ linux-2.6-2.6.21/drivers/parport/Kconfig	2007-05-13 20:58:46.000000000 +0200
@@ -9,6 +9,7 @@
 
 config PARPORT
 	tristate "Parallel port support"
+	depends on (!MAC || BROKEN)
 	---help---
 	  If you want to use devices connected to your machine's parallel port
 	  (the connector at the computer with 25 holes), e.g. printer, ZIP
@@ -34,7 +35,7 @@
 
 config PARPORT_PC
 	tristate "PC-style hardware"
-	depends on PARPORT && (!SPARC64 || PCI) && !SPARC32 && !M32R && !FRV
+	depends on PARPORT && (!SPARC64 || PCI) && !SPARC32 && !M32R && !FRV && (!MAC || BROKEN)
 	---help---
 	  You should say Y here if you have a PC-style parallel port. All
 	  IBM PC compatible computers and some Alphas have PC-style
