---
 drivers/net/atari_ethernec.c |   42 ------------------------------------------
 1 file changed, 42 deletions(-)

--- linux-m68k-2.6.21.orig/drivers/net/atari_ethernec.c
+++ linux-m68k-2.6.21/drivers/net/atari_ethernec.c
@@ -225,14 +225,6 @@ static wait_queue_head_t WaitQ;
 
 static struct delayed_work tqueue;
 
-#ifdef ETHERNEC_USE_POLL
-static struct {
-	struct work_struct poll_queue;
-	struct timer_list poll_timer;
-	struct net_device *dev;
-} poll_ops;
-#endif
-
 static struct net_device *poll_dev = NULL;
 
 static void atari_ethernec_int(struct work_struct *work)
@@ -260,22 +252,6 @@ static void atari_ethernec_int(struct wo
 		schedule_delayed_work(&tqueue, 0); /* reduced delay from 1 */
 }
 
-#ifdef ETHERNEC_USE_POLL
-static void atari_ethernec_poll_handler(unsigned long dev_addr)
-{
-	struct net_device *dev = poll_dev;
-
-	if (!dev || !dev->poll_controller)
-		return;
-
-	if (netif_running(dev))
-		dev->poll_controller(dev);
-
-	schedule_work(&poll_ops.poll_queue);
-	mod_timer(&poll_ops.poll_timer, jiffies + HZ / 100);
-}
-#endif
-
 static void atari_ethernec_start_poll(struct net_device *dev)
 {
 	poll_dev = dev;
@@ -284,19 +260,6 @@ static void atari_ethernec_start_poll(st
 
 	INIT_DELAYED_WORK(&tqueue, atari_ethernec_int);
 	schedule_delayed_work(&tqueue, 1);
-#ifdef ETHERNEC_USE_POLL
-	if (!poll_ops.poll_queue.func ||
-	    poll_ops.poll_queue.func == ei_interrupt) {
-		if (!poll_ops.poll_queue.func)
-			INIT_WORK(&poll_ops.poll_queue, ei_interrupt, dev);
-
-		init_timer(&poll_ops.poll_timer);
-		poll_ops.poll_timer.function = atari_ethernec_poll_handler;
-		poll_ops.poll_timer.expires = jiffies + HZ / 5;
-		poll_ops.poll_timer.data = (unsigned long)dev;
-		add_timer(&poll_ops.poll_timer);
-	}
-#endif
 }
 
 static void atari_ethernec_stop_poll(struct net_device *dev)
@@ -305,11 +268,6 @@ static void atari_ethernec_stop_poll(str
 
 	if (dev)
 		sleep_on(&WaitQ);
-
-#ifdef ETHERNEC_USE_POLL
-	if (poll_ops.poll_queue.func == ei_interrupt)
-		del_timer_sync(&poll_ops.poll_timer);
-#endif
 }
 
 
