From 922b48749333ddf15b9e2b55f2b0f7cf399d5e16 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Sun, 17 Jul 2011 19:48:20 +0200
Subject: [PATCH 167/290] migrate-disable-rt-variant.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 include/linux/preempt.h |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/include/linux/preempt.h b/include/linux/preempt.h
index 363e5e2..5aa7916 100644
--- a/include/linux/preempt.h
+++ b/include/linux/preempt.h
@@ -121,11 +121,15 @@ extern void migrate_enable(void);
 # define preempt_enable_rt()		preempt_enable()
 # define preempt_disable_nort()		do { } while (0)
 # define preempt_enable_nort()		do { } while (0)
+# define migrate_disable_rt()		migrate_disable()
+# define migrate_enable_rt()		migrate_enable()
 #else
 # define preempt_disable_rt()		do { } while (0)
 # define preempt_enable_rt()		do { } while (0)
 # define preempt_disable_nort()		preempt_disable()
 # define preempt_enable_nort()		preempt_enable()
+# define migrate_disable_rt()		do { } while (0)
+# define migrate_enable_rt()		do { } while (0)
 #endif
 
 #ifdef CONFIG_PREEMPT_NOTIFIERS
