From 9edeec52fd841c2807974b7db2b4ef958b4e7dba Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Mon, 14 Nov 2011 10:57:54 +0100
Subject: [187/256] rcu-more-fallout.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/rcutiny.c |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/kernel/rcutiny.c b/kernel/rcutiny.c
index 37a5444..fc581fb 100644
--- a/kernel/rcutiny.c
+++ b/kernel/rcutiny.c
@@ -368,6 +368,7 @@ void call_rcu_sched(struct rcu_head *head, void (*func)(struct rcu_head *rcu))
 }
 EXPORT_SYMBOL_GPL(call_rcu_sched);
 
+#ifndef CONFIG_PREEMPT_RT_FULL
 /*
  * Post an RCU bottom-half callback to be invoked after any subsequent
  * quiescent state.
@@ -377,3 +378,4 @@ void call_rcu_bh(struct rcu_head *head, void (*func)(struct rcu_head *rcu))
 	__call_rcu(head, func, &rcu_bh_ctrlblk);
 }
 EXPORT_SYMBOL_GPL(call_rcu_bh);
+#endif
