From 6bdd6cffbc98fdf1c6e36ad7c1dda686f2a3dc53 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 13 Jul 2011 11:03:16 +0200
Subject: [213/254] jump-label-rt.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 include/linux/jump_label.h |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/include/linux/jump_label.h b/include/linux/jump_label.h
index c513a40..f47f3e0 100644
--- a/include/linux/jump_label.h
+++ b/include/linux/jump_label.h
@@ -51,7 +51,8 @@
 #include <linux/compiler.h>
 #include <linux/workqueue.h>
 
-#if defined(CC_HAVE_ASM_GOTO) && defined(CONFIG_JUMP_LABEL)
+#if defined(CC_HAVE_ASM_GOTO) && defined(CONFIG_JUMP_LABEL) && \
+	!defined(CONFIG_PREEMPT_BASE)
 
 struct static_key {
 	atomic_t enabled;
