From 9eee911c32b17fa49c4d00541a7459ab912d4fc5 Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Mon, 16 Apr 2012 22:15:29 -0400
Subject: [PATCH 267/267] Linux 3.2.16-rt27 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..be1e37b 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt27
-- 
1.7.10

