From 776fb4d39fa4da747d2a0fd1929e3e16f4eeee92 Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Thu, 7 Jun 2012 11:22:24 -0400
Subject: [PATCH 271/271] Linux 3.2.20-rt32 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..ce6a482 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt32
-- 
1.7.10

