From 10783bd7a455b1e196d59a16d089c693ac7e8134 Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Tue, 17 Jul 2012 21:59:04 -0400
Subject: [PATCH 277/277] Linux 3.2.27-rt40 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..2af6c89 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt40
-- 
1.7.10

