From e54a80c31b40c6b0a22cd275472f9bf410d490ca Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Tue, 17 Jul 2012 21:59:04 -0400
Subject: [PATCH 278/278] Linux 3.2.28-rt42 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..8bdfb9a 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt42
-- 
1.7.10

