From ed2ad77e3f48cd12d6b0f5e55f597392827f17ca Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Mon, 17 Sep 2012 13:17:25 -0400
Subject: [PATCH 279/279] Linux 3.2.30-rt45 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..38c40b2 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt45
