From 214b9416c8681625ff8090fa0d84d2f64c126057 Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Tue, 17 Jul 2012 21:59:04 -0400
Subject: [PATCH 282/282] Linux 3.2.23-rt37 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..a3b2408 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt37
-- 
1.7.10

