From 7aefc5da2074b6de7f8ce298cfe5f206508711cb Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Tue, 16 Oct 2012 10:41:29 -0400
Subject: [PATCH 286/286] Linux 3.2.31-rt47 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..8a777ac 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt47
