From 0e157007347ba63f366d16fe31f49241bd2cfd26 Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Tue, 16 Oct 2012 10:41:29 -0400
Subject: [PATCH 287/287] Linux 3.2.32-rt48 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..2470798 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt48
