From 49349607148ad8a933c7533961a7eab3dd2f14e1 Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Fri, 2 Nov 2012 21:30:57 -0400
Subject: [PATCH 290/290] Linux 3.2.34-rt51 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..7549346 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt51
