From 67a7f182f1269da73d7e23dc844812d598e58505 Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Fri, 2 Nov 2012 21:30:57 -0400
Subject: [PATCH 290/290] Linux 3.2.35-rt52 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..66a5ed8 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt52
