From 2438ee33224bfae0ba4d969ba43bdb43447da04d Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Thu, 27 Dec 2012 10:44:19 -0500
Subject: [PATCH 293/293] Linux 3.2.37-rt55 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..51b05e9 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt55
