From 9a24c5f1ad38aeada122494a92336fd8b25283de Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Thu, 27 Dec 2012 10:44:19 -0500
Subject: [PATCH 293/293] Linux 3.2.38-rt56 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..fdb0f88 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt56
