From d4c8fcb768ea605a90ed9ad8a03dbd0d08a1b64a Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Wed, 13 Feb 2013 16:53:53 -0500
Subject: [PATCH 298/298] Linux 3.2.38-rt57 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..c06cc435 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt57
