From 680317a1d9dcefffca95f149fb75a99a0c360b96 Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Fri, 22 Feb 2013 12:08:15 -0500
Subject: [PATCH 303/303] Linux 3.2.42-rt62 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..40d81d8 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt62
