From 94798ac3f9c0475940f35b09e8ff676b1282b241 Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Fri, 22 Feb 2013 12:08:15 -0500
Subject: [PATCH 304/304] Linux 3.2.39-rt59 REBASE

---
 localversion-rt |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index b2111a2..8cdeb45 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt24
+-rt59
