Subject: arm-enable-interrupts-in-signal-code.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Sat, 16 Jul 2011 16:27:13 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 arch/arm/kernel/signal.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-3.2/arch/arm/kernel/signal.c
===================================================================
--- linux-3.2.orig/arch/arm/kernel/signal.c
+++ linux-3.2/arch/arm/kernel/signal.c
@@ -673,6 +673,9 @@ static void do_signal(struct pt_regs *re
 	if (!user_mode(regs))
 		return;
 
+	local_irq_enable();
+	preempt_check_resched();
+
 	/*
 	 * If we were from a system call, check for system call restarting...
 	 */
