From: Ingo Molnar <mingo@elte.hu>
Date: Fri, 3 Jul 2009 08:30:01 -0500
Subject: serial: 8250: Call flush_to_ldisc when the irq is threaded
Origin: https://www.kernel.org/pub/linux/kernel/projects/rt/3.12/patches-3.12.5-rt7.tar.xz

Signed-off-by: Ingo Molnar <mingo@elte.hu>

---
 drivers/tty/tty_buffer.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/tty/tty_buffer.c
+++ b/drivers/tty/tty_buffer.c
@@ -509,10 +509,14 @@ void tty_flip_buffer_push(struct tty_por
 
 	buf->tail->commit = buf->tail->used;
 
+#ifndef CONFIG_PREEMPT_RT_FULL
 	if (port->low_latency)
 		flush_to_ldisc(&buf->work);
 	else
 		schedule_work(&buf->work);
+#else
+	schedule_work(&buf->work);
+#endif
 }
 EXPORT_SYMBOL(tty_flip_buffer_push);
 
