Subject: jump-label-rt.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 13 Jul 2011 11:03:16 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 include/linux/jump_label.h |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux-3.4/include/linux/jump_label.h
===================================================================
--- linux-3.4.orig/include/linux/jump_label.h
+++ linux-3.4/include/linux/jump_label.h
@@ -51,7 +51,8 @@
 #include <linux/compiler.h>
 #include <linux/workqueue.h>
 
-#if defined(CC_HAVE_ASM_GOTO) && defined(CONFIG_JUMP_LABEL)
+#if defined(CC_HAVE_ASM_GOTO) && defined(CONFIG_JUMP_LABEL) && \
+	!defined(CONFIG_PREEMPT_BASE)
 
 struct static_key {
 	atomic_t enabled;
