Subject: mips-enable-interrupts-in-signal.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Mon, 18 Jul 2011 21:32:10 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 arch/mips/kernel/signal.c |    3 +++
 1 file changed, 3 insertions(+)

Index: linux-3.2/arch/mips/kernel/signal.c
===================================================================
--- linux-3.2.orig/arch/mips/kernel/signal.c
+++ linux-3.2/arch/mips/kernel/signal.c
@@ -604,6 +604,9 @@ static void do_signal(struct pt_regs *re
 	if (!user_mode(regs))
 		return;
 
+	local_irq_enable();
+	preempt_check_resched();
+
 	if (test_thread_flag(TIF_RESTORE_SIGMASK))
 		oldset = &current->saved_sigmask;
 	else
