Subject: mips-enable-interrupts-in-signal.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Mon, 18 Jul 2011 21:32:10 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 arch/mips/kernel/signal.c |    1 +
 1 file changed, 1 insertion(+)

Index: linux-stable/arch/mips/kernel/signal.c
===================================================================
--- linux-stable.orig/arch/mips/kernel/signal.c
+++ linux-stable/arch/mips/kernel/signal.c
@@ -606,6 +606,7 @@ asmlinkage void do_notify_resume(struct 
 	__u32 thread_info_flags)
 {
 	local_irq_enable();
+	preempt_check_resched();
 
 	/* deal with pending signal delivery */
 	if (thread_info_flags & _TIF_SIGPENDING)
