From: Thomas Gleixner <tglx@linutronix.de>
Date: Mon, 12 Aug 2013 09:16:09 +0200
Subject: mm: disable slab on rt
Origin: https://www.kernel.org/pub/linux/kernel/projects/rt/3.10/patches-3.10.10-rt7.tar.xz

---
 init/Kconfig |    1 +
 mm/slab.h    |    2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1534,6 +1534,7 @@ choice
 
 config SLAB
 	bool "SLAB"
+	depends on !PREEMPT_RT_FULL
 	help
 	  The regular slab allocator that is established and known to work
 	  well in all environments. It organizes cache hot objects in
--- a/mm/slab.h
+++ b/mm/slab.h
@@ -245,7 +245,7 @@ static inline struct kmem_cache *cache_f
  * The slab lists for all objects.
  */
 struct kmem_cache_node {
-	spinlock_t list_lock;
+	raw_spinlock_t list_lock;
 
 #ifdef CONFIG_SLAB
 	struct list_head slabs_partial;	/* partial list first, better asm code */
