Subject: net-ipv4-route-use-locks-on-up-rt.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 15 Jul 2011 16:24:45 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 net/ipv4/route.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-3.2/net/ipv4/route.c
===================================================================
--- linux-3.2.orig/net/ipv4/route.c
+++ linux-3.2/net/ipv4/route.c
@@ -251,7 +251,7 @@ struct rt_hash_bucket {
 };
 
 #if defined(CONFIG_SMP) || defined(CONFIG_DEBUG_SPINLOCK) || \
-	defined(CONFIG_PROVE_LOCKING)
+	defined(CONFIG_PROVE_LOCKING) || defined(CONFIG_PREEMPT_RT_FULL)
 /*
  * Instead of using one spinlock for each rt_hash_bucket, we use a table of spinlocks
  * The size of this table is a power of two and depends on the number of CPUS.
