From: Thomas Gleixner <tglx@linutronix.de>
Subject: panic: skip get_random_bytes for RT_FULL in init_oops_id

---
 kernel/panic.c |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-stable/kernel/panic.c
===================================================================
--- linux-stable.orig/kernel/panic.c
+++ linux-stable/kernel/panic.c
@@ -371,9 +371,11 @@ static u64 oops_id;
 
 static int init_oops_id(void)
 {
+#ifndef CONFIG_PREEMPT_RT_FULL
 	if (!oops_id)
 		get_random_bytes(&oops_id, sizeof(oops_id));
 	else
+#endif
 		oops_id++;
 
 	return 0;
