Subject: rwsem-inlcude-fix.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 15 Jul 2011 21:24:27 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 include/linux/pid.h |    1 +
 1 file changed, 1 insertion(+)

Index: linux-3.2/include/linux/pid.h
===================================================================
--- linux-3.2.orig/include/linux/pid.h
+++ linux-3.2/include/linux/pid.h
@@ -2,6 +2,7 @@
 #define _LINUX_PID_H
 
 #include <linux/rcupdate.h>
+#include <linux/atomic.h>
 
 enum pid_type
 {
