From: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
Subject: [PATCH] net/cpsw: revert 80d5c3689b886308247da295a228a54df49a44f6
Origin: https://www.kernel.org/pub/linux/kernel/projects/rt/3.10/patches-3.10.4-rt1.tar.xz

Signed-off-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
---
 drivers/net/ethernet/ti/cpsw.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-stable/drivers/net/ethernet/ti/cpsw.c
===================================================================
--- linux-stable.orig/drivers/net/ethernet/ti/cpsw.c
+++ linux-stable/drivers/net/ethernet/ti/cpsw.c
@@ -1286,7 +1286,7 @@ clean_vid:
 }
 
 static int cpsw_ndo_vlan_rx_add_vid(struct net_device *ndev,
-				    __be16 proto, u16 vid)
+		unsigned short vid)
 {
 	struct cpsw_priv *priv = netdev_priv(ndev);
 
@@ -1298,7 +1298,7 @@ static int cpsw_ndo_vlan_rx_add_vid(stru
 }
 
 static int cpsw_ndo_vlan_rx_kill_vid(struct net_device *ndev,
-				     __be16 proto, u16 vid)
+		unsigned short vid)
 {
 	struct cpsw_priv *priv = netdev_priv(ndev);
 	int ret;
