Subject: sched-disable-ttwu-queue.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Tue, 13 Sep 2011 16:42:35 +0200

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/sched_features.h |    4 ++++
 1 file changed, 4 insertions(+)

Index: linux-3.2/kernel/sched_features.h
===================================================================
--- linux-3.2.orig/kernel/sched_features.h
+++ linux-3.2/kernel/sched_features.h
@@ -60,11 +60,15 @@ SCHED_FEAT(OWNER_SPIN, 1)
  */
 SCHED_FEAT(NONTASK_POWER, 1)
 
+#ifndef CONFIG_PREEMPT_RT_FULL
 /*
  * Queue remote wakeups on the target CPU and process them
  * using the scheduler IPI. Reduces rq->lock contention/bounces.
  */
 SCHED_FEAT(TTWU_QUEUE, 1)
+#else
+SCHED_FEAT(TTWU_QUEUE, 0)
+#endif
 
 SCHED_FEAT(FORCE_SD_OVERLAP, 0)
 SCHED_FEAT(RT_RUNTIME_SHARE, 1)
