Subject: sched-clear-pf-thread-bound-on-fallback-rq.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 04 Nov 2011 20:48:36 +0100
Origin: https://www.kernel.org/pub/linux/kernel/projects/rt/3.14/patches-3.14.10-rt7.tar.xz

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/sched/core.c |    6 ++++++
 1 file changed, 6 insertions(+)

--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -1345,6 +1345,12 @@ static int select_fallback_rq(int cpu, s
 		}
 	}
 
+	/*
+	 * Clear PF_NO_SETAFFINITY, otherwise we wreckage
+	 * migrate_disable/enable. See optimization for
+	 * PF_NO_SETAFFINITY tasks there.
+	 */
+	p->flags &= ~PF_NO_SETAFFINITY;
 	return dest_cpu;
 }
 
