From: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
Date: Thu, 14 Jul 2016 14:57:07 +0200
Subject: [PATCH] sched: lazy_preempt: avoid a warning in the !RT case
Origin: https://www.kernel.org/pub/linux/kernel/projects/rt/4.6/older/patches-4.6.5-rt10.tar.xz

Signed-off-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
---
 kernel/sched/core.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -3518,7 +3518,7 @@ static __always_inline int preemptible_l
 
 #else
 
-static int preemptible_lazy(void)
+static inline int preemptible_lazy(void)
 {
 	return 1;
 }
