Subject: sched-clear-pf-thread-bound-on-fallback-rq.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 04 Nov 2011 20:48:36 +0100
Origin: https://www.kernel.org/pub/linux/kernel/projects/rt/3.10/patches-3.10.4-rt1.tar.xz

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/sched/core.c |    6 ++++++
 1 file changed, 6 insertions(+)

Index: linux-stable/kernel/sched/core.c
===================================================================
--- linux-stable.orig/kernel/sched/core.c
+++ linux-stable/kernel/sched/core.c
@@ -1245,6 +1245,12 @@ out:
 		}
 	}
 
+	/*
+	 * Clear PF_NO_SETAFFINITY, otherwise we wreckage
+	 * migrate_disable/enable. See optimization for
+	 * PF_NO_SETAFFINITY tasks there.
+	 */
+	p->flags &= ~PF_NO_SETAFFINITY;
 	return dest_cpu;
 }
 
