Subject: sysctl-include-fix.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Mon, 14 Nov 2011 10:52:34 +0100

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 include/linux/sysctl.h |    1 +
 1 file changed, 1 insertion(+)

Index: linux-3.4/include/linux/sysctl.h
===================================================================
--- linux-3.4.orig/include/linux/sysctl.h
+++ linux-3.4/include/linux/sysctl.h
@@ -933,6 +933,7 @@ enum
 #include <linux/rcupdate.h>
 #include <linux/wait.h>
 #include <linux/rbtree.h>
+#include <linux/atomic.h>
 
 /* For the /proc/sys support */
 struct ctl_table;
