Subject: x86-kvm-require-const-tsc-for-rt.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Sun, 06 Nov 2011 12:26:18 +0100

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 arch/x86/kvm/x86.c |    7 +++++++
 1 file changed, 7 insertions(+)

Index: linux-3.2/arch/x86/kvm/x86.c
===================================================================
--- linux-3.2.orig/arch/x86/kvm/x86.c
+++ linux-3.2/arch/x86/kvm/x86.c
@@ -5184,6 +5184,13 @@ int kvm_arch_init(void *opaque)
 		goto out;
 	}
 
+#ifdef CONFIG_PREEMPT_RT_FULL
+	if (!boot_cpu_has(X86_FEATURE_CONSTANT_TSC)) {
+		printk(KERN_ERR "RT requires X86_FEATURE_CONSTANT_TSC\n");
+		return -EOPNOTSUPP;
+	}
+#endif
+
 	r = kvm_mmu_module_init();
 	if (r)
 		goto out;
