--- a/arch/parisc/Makefile.orig	2006-10-12 16:44:54.000000000 +0000
+++ b/arch/parisc/Makefile	2006-10-12 16:45:21.000000000 +0000
@@ -20,8 +20,7 @@
 CHECKFLAGS	+= -D__hppa__=1
 
 ifdef CONFIG_64BIT
-CROSS_COMPILE	:= $(shell if [ -x /usr/bin/hppa64-linux-gnu-gcc ]; then \
-			echo hppa64-linux-gnu-; else echo hppa64-linux-; fi)
+CROSS_COMPILE	:= hppa64-linux-gnu-
 UTS_MACHINE	:= parisc64
 CHECKFLAGS	+= -D__LP64__=1 -m64
 else
