--- linux-2.6.17.11.ref/include/asm-ia64/mman.h	2006-08-23 23:16:33.000000000 +0200
+++ linux-2.6.17.11/include/asm-ia64/mman.h	2006-08-25 14:52:49.000000000 +0200
@@ -9,9 +9,11 @@
  */
 
 #ifdef __KERNEL__
+#if !defined(__ASSEMBLY__)
 #define arch_mmap_check	ia64_map_check_rgn
 int ia64_map_check_rgn(unsigned long addr, unsigned long len,
 		unsigned long flags);
+#endif /* !__ASSEMBLY__ */
 #endif
 
 #include <asm-generic/mman.h>
