diff -urN linux-2.6-2.6.16.old/include/linux/seccomp.h linux-2.6-2.6.16/include/linux/seccomp.h
--- linux-2.6-2.6.16.old/include/linux/seccomp.h	2006-03-20 06:53:29.000000000 +0100
+++ linux-2.6-2.6.16/include/linux/seccomp.h	2006-04-17 12:01:01.000000000 +0200
@@ -26,9 +26,12 @@
 
 #else /* CONFIG_SECCOMP */
 
 typedef struct { } seccomp_t;
+
+static inline void secure_computing(int this_syscall)
+{
+}
 
-#define secure_computing(x) do { } while (0)
 /* static inline to preserve typechecking */
 static inline int has_secure_computing(struct thread_info *ti)
 {
