From: Ben Hutchings <ben@decadent.org.uk>
Date: Thu, 13 Aug 2015 20:48:12 +0200
Subject: liblockdep: Define WRITE_ONCE(), used by <linux/rbtree_augmented.h>.

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
---
--- a/tools/lib/lockdep/uinclude/linux/compiler.h
+++ b/tools/lib/lockdep/uinclude/linux/compiler.h
@@ -4,4 +4,10 @@
 #define __used		__attribute__((__unused__))
 #define unlikely
 
+#define __ACCESS_ONCE(x) ({ \
+	__attribute__((__unused__)) typeof(x) __var = (__force typeof(x)) 0; \
+        (volatile typeof(x) *)&(x); })
+#define ACCESS_ONCE(x) (*__ACCESS_ONCE(x))
+#define WRITE_ONCE(x, val) (ACCESS_ONCE(x) = (val))
+
 #endif
