--- linux-2.6-2.6.14/kernel/audit.c.orig	2005-10-28 02:02:08.000000000 +0200
+++ linux-2.6-2.6.14/kernel/audit.c	2005-11-08 22:29:43.000000000 +0100
@@ -41,6 +41,7 @@
  * Example user-space utilities: http://people.redhat.com/sgrubb/audit/
  */
 
+#include <linux/spinlock.h>
 #include <linux/init.h>
 #include <asm/atomic.h>
 #include <asm/types.h>
--- linux-2.6-2.6.14/include/linux/dcache.h.orig	2005-10-28 02:02:08.000000000 +0200
+++ linux-2.6-2.6.14/include/linux/dcache.h	2005-11-08 22:19:45.000000000 +0100
@@ -3,9 +3,9 @@
 
 #ifdef __KERNEL__
 
+#include <linux/spinlock.h>
 #include <asm/atomic.h>
 #include <linux/list.h>
-#include <linux/spinlock.h>
 #include <linux/cache.h>
 #include <linux/rcupdate.h>
 #include <asm/bug.h>
--- linux-2.6-2.6.14/drivers/md/kcopyd.c.orig	2005-10-28 02:02:08.000000000 +0200
+++ linux-2.6-2.6.14/drivers/md/kcopyd.c	2005-11-08 22:46:17.000000000 +0100
@@ -8,6 +8,7 @@
  * completion notification.
  */
 
+#include <linux/spinlock.h>
 #include <asm/atomic.h>
 
 #include <linux/blkdev.h>
