--- source-m68k-none/arch/m68k/atari/atakeyb.c.orig	2007-01-03 19:31:17.000000000 +0100
+++ source-m68k-none/arch/m68k/atari/atakeyb.c	2007-01-03 19:35:22.000000000 +0100
@@ -662,14 +662,22 @@
     atari_turnoff_irq(IRQ_MFP_ACIA);
     do {
 	/* reset IKBD ACIA */
-	acia.key_ctrl = ACIA_RESET ; // |
-			// (atari_switches & ATARI_SWITCH_IKBD) ? ACIA_RHTID : 0;
+        if (atari_switches)
+	    acia.key_ctrl = ACIA_RESET |
+	                    (atari_switches & ATARI_SWITCH_IKBD) ? ACIA_RHTID : 0;
+        else
+            acia.key_ctrl = ACIA_RESET ;
+
 	(void)acia.key_ctrl;
 	(void)acia.key_data;
 
 	/* reset MIDI ACIA */
-	acia.mid_ctrl = ACIA_RESET ; // |
-			// (atari_switches & ATARI_SWITCH_MIDI) ? ACIA_RHTID : 0;
+        if (atari_switches)
+            acia.mid_ctrl = ACIA_RESET |
+                           (atari_switches & ATARI_SWITCH_MIDI) ? ACIA_RHTID : 0;
+        else
+            acia.mid_ctrl = ACIA_RESET ;
+
 	(void)acia.mid_ctrl;
 	(void)acia.mid_data;
 
@@ -677,12 +685,17 @@
 	/* 8 data no parity 1 start 1 stop bit */
 	/* receive interrupt enabled */
 	/* RTS low (except if switch selected), transmit interrupt disabled */
-	acia.key_ctrl = (ACIA_DIV64|ACIA_D8N1S|ACIA_RIE) ; // |
-			// ((atari_switches & ATARI_SWITCH_IKBD) ?
-			//  ACIA_RHTID : ACIA_RLTID);
-	   
-	acia.mid_ctrl = ACIA_DIV16 | ACIA_D8N1S ; // |
-			// (atari_switches & ATARI_SWITCH_MIDI) ? ACIA_RHTID : 0;
+        if (atari_switches)  {
+            acia.key_ctrl = (ACIA_DIV64|ACIA_D8N1S|ACIA_RIE) |
+                            ((atari_switches & ATARI_SWITCH_IKBD) ?
+                            ACIA_RHTID : ACIA_RLTID);
+
+            acia.mid_ctrl = ACIA_DIV16 | ACIA_D8N1S |
+                           (atari_switches & ATARI_SWITCH_MIDI) ? ACIA_RHTID : 0;
+        } else  {
+            acia.key_ctrl = (ACIA_DIV64|ACIA_D8N1S|ACIA_RIE) ;
+            acia.mid_ctrl = ACIA_DIV16 | ACIA_D8N1S ;
+        }
     }
     /* make sure the interrupt line is up */
     while ((mfp.par_dt_reg & 0x10) == 0);
