diff -urN linux-m68k/drivers/char/ioext.h linux-schmitz/drivers/char/ioext.h
--- linux-m68k/drivers/char/ioext.h	2006-11-19 21:35:33.000000000 +0100
+++ linux-schmitz/drivers/char/ioext.h	2006-11-19 21:37:26.000000000 +0100
@@ -6,7 +6,6 @@
 #ifndef _IOEXT_H_
 #define _IOEXT_H_
 
-#include <linux/config.h>
 #include <linux/netdevice.h>
 
 #include "16c552.h"
diff -urN linux-m68k/drivers/char/plip_ioext.c linux-schmitz/drivers/char/plip_ioext.c
--- linux-m68k/drivers/char/plip_ioext.c	2006-11-19 21:35:33.000000000 +0100
+++ linux-schmitz/drivers/char/plip_ioext.c	2006-11-19 21:37:26.000000000 +0100
@@ -27,7 +27,6 @@
 #include <asm/amigaints.h>
 #include <linux/zorro.h>
 
-#include <linux/config.h>
 #include <linux/kernel.h>
 #include <linux/fcntl.h>
 #include <linux/string.h>
