--- linux-orig/arch/mips/kernel/cpu-probe.c	2006-05-15 02:07:55.000000000 +0100
+++ linux-work-stable/arch/mips/kernel/cpu-probe.c	2006-05-28 23:59:12.000000000 +0100
@@ -652,7 +648,7 @@ static inline void cpu_probe_sibyte(stru
 	case PRID_IMP_SB1:
 		c->cputype = CPU_SB1;
 		/* FPU in pass1 is known to have issues. */
-		if ((c->processor_id & 0xff) < 0x20)
+		if ((c->processor_id & 0xff) < 0x02)
 			c->options &= ~(MIPS_CPU_FPU | MIPS_CPU_32FPR);
 		break;
 	case PRID_IMP_SB1A:
