This fixes a build failure caused by the devfs removal.

Signed-off-by: Thiemo Seufer <ths@networkno.de>

--- a/drivers/char/sb1250_duart.c
+++ b/drivers/char/sb1250_duart.c
@@ -763,7 +763,6 @@ static int __init sb1250_duart_init(void
 
 	sb1250_duart_driver->owner = THIS_MODULE;
 	sb1250_duart_driver->name = "ttyS";
-	sb1250_duart_driver->devfs_name = "tts/";
 	sb1250_duart_driver->major = TTY_MAJOR;
 	sb1250_duart_driver->minor_start = SB1250_DUART_MINOR_BASE;
 	sb1250_duart_driver->type            = TTY_DRIVER_TYPE_SERIAL;
