--- a/scripts/mod/modpost.c
+++ b/scripts/mod/modpost.c
@@ -16,15 +16,11 @@
 #include <ctype.h>
 #include <string.h>
 #include "modpost.h"
-#include "../../include/generated/autoconf.h"
 #include "../../include/linux/license.h"
 
 /* Some toolchains use a `_' prefix for all user symbols. */
-#ifdef CONFIG_SYMBOL_PREFIX
-#define MODULE_SYMBOL_PREFIX CONFIG_SYMBOL_PREFIX
-#else
+/* No Debian architecture currently does this. */
 #define MODULE_SYMBOL_PREFIX ""
-#endif
 
 
 /* Are we using CONFIG_MODVERSIONS? */
