From: Andrew Morton <akpm@osdl.org>

Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 drivers/char/hvc_console.c |    1 +
 1 file changed, 1 insertion(+)

diff -puN drivers/char/hvc_console.c~hvc_console-suspend-fix drivers/char/hvc_console.c
--- a/drivers/char/hvc_console.c~hvc_console-suspend-fix
+++ a/drivers/char/hvc_console.c
@@ -669,6 +669,7 @@ int khvcd(void *unused)
 	do {
 		poll_mask = 0;
 		hvc_kicked = 0;
+		try_to_freeze();
 		wmb();
 		if (cpus_empty(cpus_in_xmon)) {
 			spin_lock(&hvc_structs_lock);
_
