
From: Bastian Blank <bastian@waldi.eu.org>

Signed-off-by: Bastian Blank <bastian@waldi.eu.org>
Cc: Martin Schwidefsky <schwidefsky@de.ibm.com>
Cc: Heiko Carstens <heiko.carstens@de.ibm.com>
Cc: "David S. Miller" <davem@davemloft.net>
Cc: Jeff Garzik <jeff@garzik.org>
Acked-by: Frank Pavlic <fpavlic@de.ibm.com>
Signed-off-by: Andrew Morton <akpm@osdl.org>
---

 drivers/s390/net/qeth_main.c |    1 -
 1 files changed, 1 deletion(-)

diff -puN drivers/s390/net/qeth_main.c~s390-make-qeth-buildable drivers/s390/net/qeth_main.c
--- devel/drivers/s390/net/qeth_main.c~s390-make-qeth-buildable	2006-05-01 12:15:28.000000000 -0700
+++ devel-akpm/drivers/s390/net/qeth_main.c	2006-05-01 12:15:28.000000000 -0700
@@ -8014,7 +8014,6 @@ static int (*qeth_old_arp_constructor) (
 
 static struct neigh_ops arp_direct_ops_template = {
 	.family = AF_INET,
-	.destructor = NULL,
 	.solicit = NULL,
 	.error_report = NULL,
 	.output = dev_queue_xmit,
_

