diff -u linux-2.6.16.16-vs2.0.2-rc20/include/net/route.h linux-2.6.16.17-vs2.0.2-rc21/include/net/route.h
--- linux-2.6.16.16-vs2.0.2-rc20/include/net/route.h	2006-05-11 18:47:46 +0200
+++ linux-2.6.16.17-vs2.0.2-rc21/include/net/route.h	2006-05-11 18:47:46 +0200
@@ -229,7 +229,7 @@
 			return err;
 		if (fl.fl4_dst == IPI_LOOPBACK && !vx_check(0, VX_ADMIN))
 			fl.fl4_dst = nx_info->ipv4[0];
-#ifdef VSERVER_REMAP_SADDR
+#ifdef ONFIG_VSERVER_REMAP_SADDR
 		if (fl.fl4_src == IPI_LOOPBACK && !vx_check(0, VX_ADMIN))
 			fl.fl4_src = nx_info->ipv4[0];
 #endif
diff -u linux-2.6.16.16-vs2.0.2-rc20/net/ipv4/udp.c linux-2.6.16.17-vs2.0.2-rc21/net/ipv4/udp.c
--- linux-2.6.16.16-vs2.0.2-rc20/net/ipv4/udp.c	2006-05-11 18:49:21 +0200
+++ linux-2.6.16.17-vs2.0.2-rc21/net/ipv4/udp.c	2006-05-11 18:49:21 +0200
@@ -617,7 +617,7 @@
 				goto out;
 			if (daddr == IPI_LOOPBACK && !vx_check(0, VX_ADMIN))
 				daddr = fl.fl4_dst = nxi->ipv4[0];
-#ifdef VSERVER_REMAP_SADDR
+#ifdef CONFIG_VSERVER_REMAP_SADDR
 			if (saddr == IPI_LOOPBACK && !vx_check(0, VX_ADMIN))
 				saddr = fl.fl4_src = nxi->ipv4[0];
 #endif
