--- linux-2.6.16-vs2.0.2-rc13/fs/ioprio.c	2006-03-20 17:34:49 +0100
+++ linux-2.6.16-vs2.0.2-rc14/fs/ioprio.c	2006-03-23 19:35:17 +0100
@@ -96,7 +96,7 @@
 			if (!who)
 				user = current->user;
 			else
-				user = find_user(who, vx_current_xid());
+				user = find_user(vx_current_xid(), who);
 
 			if (!user)
 				break;
@@ -150,7 +150,7 @@
 			if (!who)
 				user = current->user;
 			else
-				user = find_user(who, vx_current_xid());
+				user = find_user(vx_current_xid(), who);
 
 			if (!user)
 				break;
--- linux-2.6.16-vs2.0.2-rc13/fs/read_write.c	2006-03-20 17:34:49 +0100
+++ linux-2.6.16/fs/read_write.c	2006-03-20 17:33:14 +0100
@@ -667,8 +667,9 @@
 		if (!(in_file->f_mode & FMODE_PREAD))
 			goto fput_in;
 	retval = rw_verify_area(READ, in_file, ppos, count);
+	if (retval < 0)
-	if (retval)
 		goto fput_in;
+	count = retval;
 
 	retval = security_file_permission (in_file, MAY_READ);
 	if (retval)
@@ -688,8 +689,9 @@
 		goto fput_out;
 	out_inode = out_file->f_dentry->d_inode;
 	retval = rw_verify_area(WRITE, out_file, &out_file->f_pos, count);
+	if (retval < 0)
-	if (retval)
 		goto fput_out;
+	count = retval;
 
 	retval = security_file_permission (out_file, MAY_WRITE);
 	if (retval)
