diff -r 99cfb47849f5 drivers/block/aoe/aoenet.c
--- a/drivers/block/aoe/aoenet.c	Mon Jul 10 13:33:16 2006 +0200
+++ b/drivers/block/aoe/aoenet.c	Tue Aug 15 14:11:32 2006 +0200
@@ -117,7 +117,7 @@ aoenet_rcv(struct sk_buff *skb, struct n
 	if (skb == NULL)
 		return 0;
 	if (skb_is_nonlinear(skb))
-	if (skb_linearize(skb, GFP_ATOMIC) < 0)
+	if (skb_linearize(skb) < 0)
 		goto exit;
 	if (!is_aoe_netif(ifp))
 		goto exit;
