diff --git a/arch/i386/kernel/fixup.c b/arch/i386/kernel/fixup.c
index 5188b23..d506a65 100644
--- a/arch/i386/kernel/fixup.c
+++ b/arch/i386/kernel/fixup.c
@@ -39,7 +39,7 @@
 
 fastcall void do_fixup_4gb_segment(struct pt_regs *regs, long error_code)
 {
-	static unsigned long printed = 0;
+	static unsigned long printed = 1;
 	char info[100];
 	int i;
 
diff --git a/arch/i386/kernel/vsyscall-note-xen.S b/arch/i386/kernel/vsyscall-note-xen.S
index c2d6dbf..019d9af 100644
--- a/arch/i386/kernel/vsyscall-note-xen.S
+++ b/arch/i386/kernel/vsyscall-note-xen.S
@@ -28,5 +28,5 @@
 #define NOTE_KERNELCAP_END ASM_ELF_NOTE_END
 
 NOTE_KERNELCAP_BEGIN(1, 1)
-NOTE_KERNELCAP(1, "nosegneg")  /* Change 1 back to 0 when glibc is fixed! */
+NOTE_KERNELCAP(0, "nosegneg")
 NOTE_KERNELCAP_END
