Upstream status: unsubmitted
Author: Troy Heber <troyh@debian.org>

diff -urpN -urN -X /space/dontdiff linux-2.6.18/arch/ia64/kernel/sal.c new/arch/ia64/kernel/sal.c
--- linux-2.6.18/arch/ia64/kernel/sal.c	2006-09-19 21:42:06.000000000 -0600
+++ new/arch/ia64/kernel/sal.c	2006-10-05 10:20:11.000000000 -0600
@@ -223,12 +223,13 @@ static void __init sal_desc_ap_wakeup(vo
  */
 static int sal_cache_flush_drops_interrupts;
 
-static void __init
+void __init
 check_sal_cache_flush (void)
 {
 	unsigned long flags;
 	int cpu;
-	u64 vector;
+	u64 vector, cache_type=3;
+	struct ia64_sal_retval isrv;
 
 	cpu = get_cpu();
 	local_irq_save(flags);
@@ -243,7 +244,7 @@ check_sal_cache_flush (void)
 	while (!ia64_get_irr(IA64_TIMER_VECTOR))
 		cpu_relax();
 
-	ia64_sal_cache_flush(3);
+	SAL_CALL(isrv, SAL_CACHE_FLUSH, cache_type, 0, 0, 0, 0, 0, 0);
 
 	if (ia64_get_irr(IA64_TIMER_VECTOR)) {
 		vector = ia64_get_ivr();
@@ -331,7 +332,6 @@ ia64_sal_init (struct ia64_sal_systab *s
 		p += SAL_DESC_SIZE(*p);
 	}
 
-	check_sal_cache_flush();
 }
 
 int
diff -urpN -urN -X /space/dontdiff linux-2.6.18/arch/ia64/kernel/setup.c new/arch/ia64/kernel/setup.c
--- linux-2.6.18/arch/ia64/kernel/setup.c	2006-09-19 21:42:06.000000000 -0600
+++ new/arch/ia64/kernel/setup.c	2006-10-05 10:15:39.000000000 -0600
@@ -458,6 +458,8 @@ setup_arch (char **cmdline_p)
 	cpu_init();	/* initialize the bootstrap CPU */
 	mmu_context_init();	/* initialize context_id bitmap */
 
+	check_sal_cache_flush();
+
 #ifdef CONFIG_ACPI
 	acpi_boot_init();
 #endif
diff -urpN -urN -X /space/dontdiff linux-2.6.18/include/asm-ia64/sal.h new/include/asm-ia64/sal.h
--- linux-2.6.18/include/asm-ia64/sal.h	2006-09-19 21:42:06.000000000 -0600
+++ new/include/asm-ia64/sal.h	2006-10-05 10:13:39.000000000 -0600
@@ -659,6 +659,7 @@ ia64_sal_freq_base (unsigned long which,
 }
 
 extern s64 ia64_sal_cache_flush (u64 cache_type);
+extern void __init check_sal_cache_flush (void);
 
 /* Initialize all the processor and platform level instruction and data caches */
 static inline s64
