diff --git a/include/linux/fs_struct.h b/include/linux/fs_struct.h
index 78a05bf..818ff7e 100644
--- a/include/linux/fs_struct.h
+++ b/include/linux/fs_struct.h
@@ -4,11 +4,14 @@
 #include <linux/path.h>
 
 struct fs_struct {
-	int users;
+	atomic_t count;
 	rwlock_t lock;
 	int umask;
-	int in_exec;
 	struct path root, pwd;
+#ifndef __GENKSYMS__
+	int users;
+	int in_exec;
+#endif
 };
 
 extern struct kmem_cache *fs_cachep;
