diff -Naurp linux-2.6.orig/fs/namei.c linux-2.6/fs/namei.c
--- linux-2.6.orig/fs/namei.c	2008-01-24 22:58:37.000000000 +0000
+++ linux-2.6/fs/namei.c	2008-02-11 19:53:38.000000000 +0000
@@ -389,6 +389,7 @@ void release_open_intent(struct nameidat
 	else
 		fput(nd->intent.open.file);
 }
+EXPORT_SYMBOL_GPL(release_open_intent);
 
 static inline struct dentry *
 do_revalidate(struct dentry *dentry, struct nameidata *nd)
