diff -urN ../orig/linux-2.6-2.6.13+2.6.14-rc5/arch/ppc/Makefile linux-2.6-2.6.13+2.6.14-rc5/arch/ppc/Makefile
--- ../orig/linux-2.6-2.6.13+2.6.14-rc5/arch/ppc/Makefile	2005-10-22 18:17:48.000000000 +0000
+++ linux-2.6-2.6.13+2.6.14-rc5/arch/ppc/Makefile	2005-10-23 13:27:47.000000000 +0000
@@ -113,7 +113,7 @@
 include/asm: arch/$(ARCH)/include/asm
 arch/$(ARCH)/include/asm:
 	$(Q)if [ ! -d arch/$(ARCH)/include ]; then mkdir -p arch/$(ARCH)/include; fi
-	$(Q)ln -fsn $(srctree)/include/asm-powerpc arch/$(ARCH)/include/asm
+	$(Q)ln -fsn $(if $(KBUILD_SRC),$(KBUILD_SRC),../../..)/include/asm-powerpc arch/$(ARCH)/include/asm
 
 # Use the file '.tmp_gas_check' for binutils tests, as gas won't output
 # to stdout and these checks are run even on install targets.
