diff -aurN a/drivers/scsi/Kconfig b/drivers/scsi/Kconfig
--- a/drivers/scsi/Kconfig	2005-06-17 15:48:29.000000000 -0400
+++ b/drivers/scsi/Kconfig	2005-06-18 15:19:35.000000000 -0400
@@ -1312,8 +1312,6 @@
 	  To compile this driver as a module, choose M here: the
 	  module will be called qlogicpti.
 
-source "drivers/scsi/qla2xxx/Kconfig"
-
 config SCSI_LPFC
 	tristate "Emulex LightPulse Fibre Channel Support"
 	depends on PCI && SCSI
diff -aurN a/drivers/scsi/Makefile b/drivers/scsi/Makefile
--- a/drivers/scsi/Makefile	2005-06-17 15:48:29.000000000 -0400
+++ b/drivers/scsi/Makefile	2005-06-18 15:19:15.000000000 -0400
@@ -79,7 +79,6 @@
 obj-$(CONFIG_SCSI_QLOGIC_ISP)	+= qlogicisp.o 
 obj-$(CONFIG_SCSI_QLOGIC_FC)	+= qlogicfc.o 
 obj-$(CONFIG_SCSI_QLOGIC_1280)	+= qla1280.o 
-obj-$(CONFIG_SCSI_QLA2XXX)	+= qla2xxx/
 obj-$(CONFIG_SCSI_LPFC)		+= lpfc/
 obj-$(CONFIG_SCSI_PAS16)	+= pas16.o
 obj-$(CONFIG_SCSI_SEAGATE)	+= seagate.o
