aufs2.1 kbuild patch for linux-2.6.37

diff --git a/fs/Kconfig b/fs/Kconfig
index 771f457..bb1a52f 100644
--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -191,6 +191,7 @@ source "fs/romfs/Kconfig"
 source "fs/sysv/Kconfig"
 source "fs/ufs/Kconfig"
 source "fs/exofs/Kconfig"
+source "fs/aufs/Kconfig"
 
 endif # MISC_FILESYSTEMS
 
diff --git a/fs/Makefile b/fs/Makefile
index a7f7cef..95dd4d3 100644
--- a/fs/Makefile
+++ b/fs/Makefile
@@ -121,3 +121,4 @@ obj-$(CONFIG_BTRFS_FS)		+= btrfs/
 obj-$(CONFIG_GFS2_FS)           += gfs2/
 obj-$(CONFIG_EXOFS_FS)          += exofs/
 obj-$(CONFIG_CEPH_FS)		+= ceph/
+obj-$(CONFIG_AUFS_FS)           += aufs/
diff --git a/include/linux/Kbuild b/include/linux/Kbuild
index 97319a8..7ebb4b4 100644
--- a/include/linux/Kbuild
+++ b/include/linux/Kbuild
@@ -60,6 +60,7 @@ header-y += atmppp.h
 header-y += atmsap.h
 header-y += atmsvc.h
 header-y += audit.h
+header-y += aufs_type.h
 header-y += auto_fs.h
 header-y += auto_fs4.h
 header-y += auxvec.h
