From: Ben Hutchings <ben@decadent.org.uk>
Subject: [PATCH] aufs: mark as staging

I really don't want to support this.

--- a/fs/aufs/module.c
+++ b/fs/aufs/module.c
@@ -81,6 +81,7 @@
 MODULE_DESCRIPTION(AUFS_NAME
 	" -- Advanced multi layered unification filesystem");
 MODULE_VERSION(AUFS_VERSION);
+MODULE_INFO(staging, "Y");
 
 /* it should be 'byte', but param_set_byte() prints it by "%c" */
 short aufs_nwkq = AUFS_NWKQ_DEF;
