--- a/include/asm-alpha/io.h	2007-04-25 20:08:32.000000000 -0700
+++ b/include/asm-alpha/io.h	2007-05-16 19:01:44.000000000 -0700
@@ -89,6 +89,7 @@
 #endif
 
 #define page_to_phys(page)	page_to_pa(page)
+#define isa_page_to_bus page_to_phys
 
 /* This depends on working iommu.  */
 #define BIO_VMERGE_BOUNDARY	(alpha_mv.mv_pci_tbi ? PAGE_SIZE : 0)
@@ -126,6 +127,7 @@
 	virt = phys_to_virt(address);
 	return (long)address <= 0 ? NULL : virt;
 }
+#define isa_bus_to_virt bus_to_virt
 
 /*
  * There are different chipsets to interface the Alpha CPUs to the world.
