diff --git a/drivers/xen/console/console.c b/drivers/xen/console/console.c
index e4e5d41..03f8ce9 100644
--- a/drivers/xen/console/console.c
+++ b/drivers/xen/console/console.c
@@ -66,20 +66,25 @@
  *  'xencons=tty'  [XC_TTY]:     Console attached to '/dev/tty[0-9]+'.
  *  'xencons=ttyS' [XC_SERIAL]:  Console attached to '/dev/ttyS[0-9]+'.
  *  'xencons=xvc'  [XC_XVC]:     Console attached to '/dev/xvc0'.
- *  default:                     XC_XVC
+ *  'xencons=hvc'  [XC_HVC]:     Console attached to '/dev/hvc0'.
+ *  default:                     XC_HVC
  * 
  * NB. In mode XC_TTY, we create dummy consoles for tty2-63. This suppresses
  * warnings from standard distro startup scripts.
  */
 static enum {
-	XC_OFF, XC_TTY, XC_SERIAL, XC_XVC
-} xc_mode = XC_XVC;
+	XC_OFF, XC_TTY, XC_SERIAL, XC_XVC, XC_HVC
+} xc_mode = XC_HVC;
 static int xc_num = -1;
 
 /* /dev/xvc0 device number allocated by lanana.org. */
 #define XEN_XVC_MAJOR 204
 #define XEN_XVC_MINOR 191
 
+/* /dev/hvc0 device number */
+#define XEN_HVC_MAJOR 229
+#define XEN_HVC_MINOR 0
+
 static int __init xencons_setup(char *str)
 {
 	char *q;
@@ -97,6 +102,9 @@ static int __init xencons_setup(char *str)
 	} else if (!strncmp(str, "xvc", 3)) {
 		xc_mode = XC_XVC;
 		str += 3;
+	} else if (!strncmp(str, "hvc", 3)) {
+		xc_mode = XC_HVC;
+		str += 3;
 	} else if (!strncmp(str, "off", 3)) {
 		xc_mode = XC_OFF;
 		str += 3;
@@ -205,6 +213,14 @@ static int __init xen_console_init(void)
 			xc_num = 0;
 		break;
 
+	case XC_HVC:
+		strcpy(kcons_info.name, "hvc");
+		if (xc_num == -1)
+			xc_num = 0;
+		if (!is_initial_xendomain())
+			add_preferred_console(kcons_info.name, xc_num, NULL);
+		break;
+
 	case XC_SERIAL:
 		strcpy(kcons_info.name, "ttyS");
 		if (xc_num == -1)
@@ -681,6 +697,12 @@ static int __init xencons_init(void)
 		DRV(xencons_driver)->minor_start = XEN_XVC_MINOR;
 		DRV(xencons_driver)->name_base   = xc_num;
 		break;
+	case XC_HVC:
+		DRV(xencons_driver)->name        = "hvc";
+		DRV(xencons_driver)->major       = XEN_HVC_MAJOR;
+		DRV(xencons_driver)->minor_start = XEN_HVC_MINOR;
+		DRV(xencons_driver)->name_base   = xc_num;
+		break;
 	case XC_SERIAL:
 		DRV(xencons_driver)->name        = "ttyS";
 		DRV(xencons_driver)->minor_start = 64 + xc_num;
