From: Ian Campbell <ijc@hellion.org.uk>
Date: Wed, 16 Jan 2013 13:38:20 +0000
Subject: microcode_xen: update for changed core API
Forwarded: no

Signed-off-by: Ian Campbell <ijc@hellion.org.uk>
---
 arch/x86/kernel/microcode_xen.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/arch/x86/kernel/microcode_xen.c b/arch/x86/kernel/microcode_xen.c
index 9e50566..6b7aa48 100644
--- a/arch/x86/kernel/microcode_xen.c
+++ b/arch/x86/kernel/microcode_xen.c
@@ -56,7 +56,8 @@ static int xen_microcode_update(int cpu)
 	return err;
 }
 
-static enum ucode_state xen_request_microcode_fw(int cpu, struct device *device)
+static enum ucode_state xen_request_microcode_fw(int cpu, struct device *device,
+						 bool refresh_fw)
 {
 	char name[36];
 	struct cpuinfo_x86 *c = &cpu_data(cpu);
-- 
1.7.2.5

