diff --git a/arch/x86/kernel/Makefile b/arch/x86/kernel/Makefile
index 81f7529..c41a5c8 100644
--- a/arch/x86/kernel/Makefile
+++ b/arch/x86/kernel/Makefile
@@ -91,8 +91,6 @@ scx200-y			+= scx200_32.o
 
 obj-$(CONFIG_OLPC)		+= olpc.o
 
-obj-$(CONFIG_X86_XEN)		+= fixup.o
-
 ###
 # 64 bit specific files
 ifeq ($(CONFIG_X86_64),y)
diff --git a/arch/x86/kernel/entry_32-xen.S b/arch/x86/kernel/entry_32-xen.S
index ca66938..37f6f70 100644
--- a/arch/x86/kernel/entry_32-xen.S
+++ b/arch/x86/kernel/entry_32-xen.S
@@ -1238,15 +1238,8 @@ ENTRY(spurious_interrupt_bug)
 	CFI_ADJUST_CFA_OFFSET 4
 	jmp error_code
 	CFI_ENDPROC
-#endif /* !CONFIG_XEN */
-
-ENTRY(fixup_4gb_segment)
-	RING0_EC_FRAME
-	pushl $do_fixup_4gb_segment
-	CFI_ADJUST_CFA_OFFSET 4
-	jmp error_code
-	CFI_ENDPROC
 END(spurious_interrupt_bug)
+#endif /* !CONFIG_XEN */
 
 ENTRY(kernel_thread_helper)
 	pushl $0		# fake return address for unwinder
diff --git a/arch/x86/kernel/traps_32-xen.c b/arch/x86/kernel/traps_32-xen.c
index c174679..0ce86ca 100644
--- a/arch/x86/kernel/traps_32-xen.c
+++ b/arch/x86/kernel/traps_32-xen.c
@@ -101,8 +101,6 @@ asmlinkage void simd_coprocessor_error(void);
 asmlinkage void alignment_check(void);
 #ifndef CONFIG_XEN
 asmlinkage void spurious_interrupt_bug(void);
-#else
-asmlinkage void fixup_4gb_segment(void);
 #endif
 asmlinkage void machine_check(void);
 
@@ -1172,7 +1170,6 @@ static const trap_info_t __cpuinitconst trap_table[] = {
 	{ 12, 0, __KERNEL_CS, (unsigned long)stack_segment		},
 	{ 13, 0, __KERNEL_CS, (unsigned long)general_protection		},
 	{ 14, 0|4, __KERNEL_CS, (unsigned long)page_fault		},
-	{ 15, 0, __KERNEL_CS, (unsigned long)fixup_4gb_segment		},
 	{ 16, 0, __KERNEL_CS, (unsigned long)coprocessor_error		},
 	{ 17, 0, __KERNEL_CS, (unsigned long)alignment_check		},
 #ifdef CONFIG_X86_MCE
