diff -urN test1/net/core/dev.c test2/net/core/dev.c
--- test1/net/core/dev.c	2006-09-25 21:01:06.000000000 +0200
+++ test2/net/core/dev.c	2006-09-25 21:01:44.000000000 +0200
@@ -113,11 +113,11 @@
 #include <linux/wireless.h>
 #include <net/iw_handler.h>
 #include <asm/current.h>
+#include <linux/vs_network.h>
 #include <linux/audit.h>
 #include <linux/dmaengine.h>
 #include <linux/err.h>
 #include <linux/ctype.h>
-#include <linux/vs_network.h>
 
 /*
  *	The list of packet types we will receive (as opposed to discard)
