From: H.J. Lu <hjl.tools@gmail.com>
Date: Sat, 26 Feb 2011 20:37:08 +0000 (-0800)
Subject: Correct a typo in xen_do_hypervisor_callback label.

Correct a typo in xen_do_hypervisor_callback label.
---
diff --git a/arch/x86/kernel/entry_64.S b/arch/x86/kernel/entry_64.S
index 44a99bb..3571494 100644
--- a/arch/x86/kernel/entry_64.S
+++ b/arch/x86/kernel/entry_64.S
@@ -1270,7 +1270,7 @@ ENTRY(xen_do_hypervisor_callback)   # do_hypervisor_callback(struct *pt_regs)
 	decl PER_CPU_VAR(irq_count)
 	jmp  error_exit
 	CFI_ENDPROC
-END(do_hypervisor_callback)
+END(xen_do_hypervisor_callback)
 
 /*
  * Hypervisor uses this for application faults while it executes.
