From: Stefan Hajnoczi <stefanha@gmail.com>
Date: Wed, 14 Nov 2012 08:50:43 +0100
Subject: [057/106] staging: line6: drop unused CONFIG_LINE6_USB_DUMP_CTRL

commit 4aea449ac8c4955f18c49ecb6fd3e0417a2073c0 upstream.

Signed-off-by: Stefan Hajnoczi <stefanha@gmail.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 drivers/staging/line6/Kconfig |    9 ---------
 1 file changed, 9 deletions(-)

diff --git a/drivers/staging/line6/Kconfig b/drivers/staging/line6/Kconfig
index a5ded12..2101799 100644
--- a/drivers/staging/line6/Kconfig
+++ b/drivers/staging/line6/Kconfig
@@ -23,15 +23,6 @@ menuconfig LINE6_USB
 
 if LINE6_USB
 
-config LINE6_USB_DUMP_CTRL
-	bool "dump control messages"
-	default n
-	help
-	  Say Y here to write control messages sent to and received from
-	  Line6 devices to the syslog.
-
-	  If unsure, say N.
-
 config LINE6_USB_DUMP_MIDI
 	bool "dump MIDI messages"
 	default n
