--- linux-2.6.16.43-vs2.0.3-rc1/fs/locks.c	2007-03-28 02:23:15 +0200
+++ linux-2.6.16.43-vs2.0.3-rc2/fs/locks.c	2007-03-17 16:20:04 +0100
@@ -759,6 +759,7 @@ static int flock_lock_file(struct file *
 	new_fl = locks_alloc_lock();
 	if (new_fl == NULL)
 		goto out;
+	new_fl->fl_xid = -1;
 	/*
 	 * If a higher-priority process was blocked on the old file lock,
 	 * give it the opportunity to lock the file.
@@ -780,8 +781,8 @@ static int flock_lock_file(struct file *
	if (request->fl_flags & FL_ACCESS)
 		goto out;
 	locks_copy_lock(new_fl, request);
-	vx_locks_inc(new_fl);
 	locks_insert_lock(&inode->i_flock, new_fl);
+	vx_locks_inc(new_fl);
 	new_fl = NULL;
 	error = 0;
 
@@ -1383,8 +1384,8 @@ static int __setlease(struct file *filp,
 		goto out;
 
 	locks_copy_lock(fl, lease);
-
 	locks_insert_lock(before, fl);
+	vx_locks_inc(fl);
 
 	*flp = fl;
 	error = 0;
