diff -ruN a/drivers/net/Kconfig b/drivers/net/Kconfig
--- a/drivers/net/Kconfig	2007-02-04 19:44:54.000000000 +0100
+++ b/drivers/net/Kconfig	2007-02-25 01:47:11.000000000 +0100
@@ -1438,7 +1438,7 @@
 
 config FORCEDETH_NAPI
 	bool "Use Rx and Tx Polling (NAPI) (EXPERIMENTAL)"
-	depends on FORCEDETH && EXPERIMENTAL
+	depends on FORCEDETH && EXPERIMENTAL && BROKEN
 	help
 	  NAPI is a new driver API designed to reduce CPU and interrupt load
 	  when the driver is receiving lots of packets from the card. It is
