diff --git a/Makefile b/Makefile
index 0e9eee7..4ea4898 100644
--- a/Makefile
+++ b/Makefile
@@ -807,7 +807,7 @@ space      := $(nullstring) # end of line
 ___localver = $(objtree)/localversion* $(srctree)/localversion*
 __localver  = $(sort $(wildcard $(___localver)))
 # skip backup files (containing '~')
-_localver = $(foreach f, $(__localver), $(if $(findstring ~, $(f)),,$(f)))
+_localver = $(foreach f, $(__localver), $(if $(findstring ~, $(notdir $(f))),,$(f)))
 
 localver = $(subst $(space),, \
 	   $(shell cat /dev/null $(_localver)) \
