From netdev-owner@vger.kernel.org Fri Sep 22 15:58:36 2006
From: Francois Romieu <romieu@fr.zoreil.com>

    r8169: the MMIO region of the 8167 stands behin BAR#1
     

diff --git a/drivers/net/r8169.c b/drivers/net/r8169.c
index 805562b..93cd1f4 100644
--- a/drivers/net/r8169.c
+++ b/drivers/net/r8169.c
@@ -210,7 +210,7 @@ static const struct {
 static struct pci_device_id rtl8169_pci_tbl[] = {
 	{ PCI_DEVICE(PCI_VENDOR_ID_REALTEK,	0x8129), 0, 0, RTL_CFG_0 },
 	{ PCI_DEVICE(PCI_VENDOR_ID_REALTEK,	0x8136), 0, 0, RTL_CFG_2 },
-	{ PCI_DEVICE(PCI_VENDOR_ID_REALTEK,	0x8167), 0, 0, RTL_CFG_1 },
+	{ PCI_DEVICE(PCI_VENDOR_ID_REALTEK,	0x8167), 0, 0, RTL_CFG_0 },
 	{ PCI_DEVICE(PCI_VENDOR_ID_REALTEK,	0x8168), 0, 0, RTL_CFG_2 },
 	{ PCI_DEVICE(PCI_VENDOR_ID_REALTEK,	0x8169), 0, 0, RTL_CFG_0 },
 	{ PCI_DEVICE(PCI_VENDOR_ID_DLINK,	0x4300), 0, 0, RTL_CFG_0 },

-- 
Ueimor

