From netdev-owner@vger.kernel.org Fri Oct 13 07:35:57 2006
Date: Thu, 12 Oct 2006 22:30:46 +0200
From: Francois Romieu <romieu@fr.zoreil.com>

Arnaud Patard:
      r8169: fix infinite loop during hotplug

diff --git a/drivers/net/r8169.c b/drivers/net/r8169.c
index 4c47c5b..c2c9a86 100644
--- a/drivers/net/r8169.c
+++ b/drivers/net/r8169.c
@@ -2701,6 +2702,7 @@ static void rtl8169_down(struct net_devi
 	struct rtl8169_private *tp = netdev_priv(dev);
 	void __iomem *ioaddr = tp->mmio_addr;
 	unsigned int poll_locked = 0;
+	unsigned int intrmask;
 
 	rtl8169_delete_timer(dev);
 
@@ -2739,8 +2741,11 @@ core_down:
 	 * 2) dev->change_mtu
 	 *    -> rtl8169_poll can not be issued again and re-enable the
 	 *       interruptions. Let's simply issue the IRQ down sequence again.
+	 *
+	 * No loop if hotpluged or major error (0xffff).
 	 */
-	if (RTL_R16(IntrMask))
+	intrmask = RTL_R16(IntrMask);
+	if (intrmask && (intrmask != 0xffff))
 		goto core_down;
 
 	rtl8169_tx_clear(tp);
-- 
Ueimor


