diff -aur a/drivers/net/sunhme.c b/drivers/net/sunhme.c
--- a/drivers/net/sunhme.c	2006-11-18 19:28:22.000000000 -0800
+++ b/drivers/net/sunhme.c	2006-11-23 10:36:36.000000000 -0800
@@ -3012,6 +3012,11 @@
 #endif
 
 	err = -ENODEV;
+
+	if (pci_enable_device(pdev))
+		goto err_out;
+	pci_set_master(pdev);
+
 	if (!strcmp(prom_name, "SUNW,qfe") || !strcmp(prom_name, "qfe")) {
 		qp = quattro_pci_find(pdev);
 		if (qp == NULL)
