diff -urpN linux-source-2.6.29.orig/arch/sparc/boot/Makefile linux-source-2.6.29/arch/sparc/boot/Makefile
--- linux-source-2.6.29.orig/arch/sparc/boot/Makefile	2009-03-23 17:12:14.000000000 -0600
+++ linux-source-2.6.29/arch/sparc/boot/Makefile	2009-05-15 16:14:09.000000000 -0600
@@ -79,6 +79,9 @@ $(obj)/image: vmlinux FORCE
 	$(call if_changed,strip)
 	@echo '  kernel: $@ is ready'
 
+$(obj)/zImage: $(obj)/image
+	$(call if_changed,gzip)
+
 $(obj)/tftpboot.img: vmlinux $(obj)/piggyback_64 System.map $(ROOT_IMG) FORCE
 	$(call if_changed,elftoaout)
 	$(call if_changed,piggy)
diff -urpN linux-source-2.6.29.orig/arch/sparc/Makefile linux-source-2.6.29/arch/sparc/Makefile
--- linux-source-2.6.29.orig/arch/sparc/Makefile	2009-03-23 17:12:14.000000000 -0600
+++ linux-source-2.6.29/arch/sparc/Makefile	2009-05-15 16:14:44.000000000 -0600
@@ -91,6 +91,8 @@ endif
 
 boot := arch/sparc/boot
 
+all: zImage
+
 image zImage tftpboot.img vmlinux.aout: vmlinux
 	$(Q)$(MAKE) $(build)=$(boot) $(boot)/$@
 
