Subject: scripts/recordmcount.pl: There is no -m32 option on Super-H
From: Michael Karcher <kernel@mkarcher.dialup.fu-berlin.de>
Date: Sat, 17 Jan 2015 23:36:15 +0000
Origin: http://marc.info/?l=linux-kernel&m=142153803030712&w=2

Signed-Off-By: Michael Karcher <kernel@mkarcher.dialup.fu-berlin.de>
---
 scripts/recordmcount.pl | 1 -
 1 file changed, 1 deletion(-)

diff --git a/scripts/recordmcount.pl b/scripts/recordmcount.pl
index 56ea99a..537c38c 100755
--- a/scripts/recordmcount.pl
+++ b/scripts/recordmcount.pl
@@ -255,7 +255,6 @@ if ($arch eq "x86_64") {
     # force flags for this arch
     $ld .= " -m shlelf_linux";
     $objcopy .= " -O elf32-sh-linux";
-    $cc .= " -m32";
 
 } elsif ($arch eq "powerpc") {
     $local_regex = "^[0-9a-fA-F]+\\s+t\\s+(\\.?\\S+)";
