diff -aur a/drivers/video/sunxvr500.c b/drivers/video/sunxvr500.c
--- a/drivers/video/sunxvr500.c	2008-11-20 18:02:37.000000000 -0500
+++ b/drivers/video/sunxvr500.c	2008-12-01 15:43:04.000000000 -0500
@@ -396,6 +396,7 @@
 
 static struct pci_device_id e3d_pci_table[] = {
 	{	PCI_DEVICE(PCI_VENDOR_ID_3DLABS, 0x7a0),	},
+	{	PCI_DEVICE(0x1091, 0x7a0),			},
 	{	PCI_DEVICE(PCI_VENDOR_ID_3DLABS, 0x7a2),	},
 	{	.vendor = PCI_VENDOR_ID_3DLABS,
 		.device = PCI_ANY_ID,
