diff -aur a/drivers/video/sunxvr500.c b/drivers/video/sunxvr500.c
--- a/drivers/video/sunxvr500.c	2008-11-20 18:02:37.000000000 -0500
+++ b/drivers/video/sunxvr500.c	2008-12-01 15:43:04.000000000 -0500
@@ -396,6 +396,7 @@
 
 static struct pci_device_id e3d_pci_table[] = {
 	{	PCI_DEVICE(PCI_VENDOR_ID_3DLABS, 0x7a0),	},
+	{	PCI_DEVICE(PCI_VENDOR_ID_IGRAPH, 0x7a0),	},
 	{	PCI_DEVICE(PCI_VENDOR_ID_3DLABS, 0x7a2),	},
 	{	.vendor = PCI_VENDOR_ID_3DLABS,
 		.device = PCI_ANY_ID,
diff -aur a/include/linux/pci_ids.h b/include/linux/pci_ids.h
--- a/include/linux/pci_ids.h	2008-11-20 18:02:37.000000000 -0500
+++ b/include/linux/pci_ids.h	2008-12-01 15:43:04.000000000 -0500
@@ -930,6 +930,8 @@
 #define PCI_DEVICE_ID_SUN_TOMATILLO	0xa801
 #define PCI_DEVICE_ID_SUN_CASSINI	0xabba
 
+#define PCI_VENDOR_ID_IGRAPH		0x1091
+
 #define PCI_VENDOR_ID_CMD		0x1095
 #define PCI_DEVICE_ID_CMD_643		0x0643
 #define PCI_DEVICE_ID_CMD_646		0x0646
