Needed otherwise you get a 3 GB kernel file

From: Lennert Buytenhek <buytenh@wantstofly.org>
Subject: Re: What's queued for 2.6.24
Date: Wed, 3 Oct 2007 16:12:06 +0200

diff -ur linux-2.6.23-rc4.orig/arch/arm/Makefile linux-2.6.23-rc4/arch/arm/Makefile
--- linux-2.6.23-rc4.orig/arch/arm/Makefile	2007-08-28 11:07:38.000000000 +0200
+++ linux-2.6.23-rc4/arch/arm/Makefile	2007-10-03 16:09:12.000000000 +0200
@@ -12,7 +12,7 @@
 
 LDFLAGS_vmlinux	:=-p --no-undefined -X
 CPPFLAGS_vmlinux.lds = -DTEXT_OFFSET=$(TEXT_OFFSET)
-OBJCOPYFLAGS	:=-O binary -R .note -R .comment -S
+OBJCOPYFLAGS	:=-O binary -R .note -R .note.gnu.build-id -R .comment -S
 GZFLAGS		:=-9
 #CFLAGS		+=-pipe
 # Explicitly specifiy 32-bit ARM ISA since toolchain default can be -mthumb:

