diff --git a/drivers/net/Kconfig b/drivers/net/Kconfig
index e55c1c0..e706744 100644
--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -2616,7 +2616,7 @@ config MYRI10GE
 
 config NETXEN_NIC
 	tristate "NetXen Multi port (1/10) Gigabit Ethernet NIC"
-	depends on PCI && (!SPARC32 || BROKEN)
+	depends on PCI && (!SPARC32 || BROKEN) && (!ARM || BROKEN)
 	help
 	  This enables the support for NetXen's Gigabit Ethernet card.
 
