http://bugzilla.kernel.org/show_bug.cgi?id=11334

diff --git a/drivers/net/Kconfig b/drivers/net/Kconfig
index a5c141c..2f25dfc 100644
--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -2428,7 +2428,7 @@ config S2IO
 
 config MYRI10GE
 	tristate "Myricom Myri-10G Ethernet support"
-	depends on PCI && INET
+	depends on PCI && INET && (!ARM || BROKEN)
 	select FW_LOADER
 	select CRC32
 	select INET_LRO
