diff --git a/drivers/net/Kconfig b/drivers/net/Kconfig
index c53c9b6..1ed5527 100644
--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -2488,7 +2488,7 @@ config MYRI10GE
 
 config NETXEN_NIC
 	tristate "NetXen Multi port (1/10) Gigabit Ethernet NIC"
-	depends on PCI && (!SPARC32 || BROKEN)
+	depends on PCI && (!SPARC32 || BROKEN) && (!ARM || BROKEN)
 	help
 	  This enables the support for NetXen's Gigabit Ethernet card.
 
