--- a/drivers/net/wireless/bcm43xx/Kconfig	2007-09-27 12:50:43.000000000 +0000
+++ b/drivers/net/wireless/bcm43xx/Kconfig	2007-09-27 12:51:08.000000000 +0000
@@ -1,6 +1,6 @@
 config BCM43XX
 	tristate "Broadcom BCM43xx wireless support"
-	depends on PCI && IEEE80211 && IEEE80211_SOFTMAC && WLAN_80211 && EXPERIMENTAL
+	depends on PCI && IEEE80211 && IEEE80211_SOFTMAC && WLAN_80211 && EXPERIMENTAL && (!ARM || BROKEN)
 	select WIRELESS_EXT
 	select FW_LOADER
 	select HW_RANDOM
