--- a/drivers/media/video/bt8xx/Kconfig	2008-05-19 19:06:35.000000000 +0000
+++ b/drivers/media/video/bt8xx/Kconfig	2008-05-19 19:07:17.000000000 +0000
@@ -1,6 +1,6 @@
 config VIDEO_BT848
 	tristate "BT848 Video For Linux"
-	depends on VIDEO_DEV && PCI && I2C && VIDEO_V4L2 && INPUT
+	depends on VIDEO_DEV && PCI && I2C && VIDEO_V4L2 && INPUT && (!ARM || BROKEN)
 	depends on HOTPLUG	# due to FW_LOADER
 	select I2C_ALGOBIT
 	select FW_LOADER

