diff -Naur linux-2.6-2.6.20.orig/drivers/mtd/nand/Kconfig linux-2.6-2.6.20/drivers/mtd/nand/Kconfig
--- linux-2.6-2.6.20.orig/drivers/mtd/nand/Kconfig	2007-02-04 11:44:54.000000000 -0700
+++ linux-2.6-2.6.20/drivers/mtd/nand/Kconfig	2007-04-18 22:02:18.000000000 -0600
@@ -223,7 +223,7 @@
 
 config MTD_NAND_CAFE
        tristate "NAND support for OLPC CAFÉ chip"
-       depends on MTD_NAND && PCI
+       depends on MTD_NAND && PCI && (!ARM || BROKEN)
        help
 	 Use NAND flash attached to the CAFÉ chip designed for the $100
 	 laptop.
diff -Naur linux-2.6-2.6.20.orig/drivers/net/Kconfig linux-2.6-2.6.20/drivers/net/Kconfig
--- linux-2.6-2.6.20.orig/drivers/net/Kconfig	2007-04-18 22:16:48.000000000 -0600
+++ linux-2.6-2.6.20/drivers/net/Kconfig	2007-04-18 22:16:29.000000000 -0600
@@ -2484,7 +2484,7 @@
 
 config NETXEN_NIC
 	tristate "NetXen Multi port (1/10) Gigabit Ethernet NIC"
-	depends on PCI && (!SPARC32 || BROKEN)
+	depends on PCI && (!SPARC32 || BROKEN) && (!ARM || BROKEN)
 	help
 	  This enables the support for NetXen's Gigabit Ethernet card.
 
