diff --git a/drivers/net/Kconfig b/drivers/net/Kconfig
index 81ef81c..c53c9b6 100644
--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -2379,7 +2379,7 @@ config CHELSIO_T1_NAPI
 
 config CHELSIO_T3
 	tristate "Chelsio Communications T3 10Gb Ethernet support"
-	depends on PCI
+	depends on PCI && (!ARM || BROKEN)
 	select FW_LOADER
 	help
 	  This driver supports Chelsio T3-based gigabit and 10Gb Ethernet
