--- linux-2.6-2.6.20/drivers/net/Kconfig~	2007-04-25 14:10:24.000000000 +0000
+++ linux-2.6-2.6.20/drivers/net/Kconfig	2007-04-25 14:11:20.000000000 +0000
@@ -2373,7 +2373,7 @@
 
 config CHELSIO_T3
 	tristate "Chelsio Communications T3 10Gb Ethernet support"
-	depends on PCI
+	depends on PCI && (!ARM || BROKEN)
 	select FW_LOADER
 	help
 	  This driver supports Chelsio T3-based gigabit and 10Gb Ethernet
