diff --git a/drivers/watchdog/Kconfig b/drivers/watchdog/Kconfig
index 09a3d55..380409f 100644
--- a/drivers/watchdog/Kconfig
+++ b/drivers/watchdog/Kconfig
@@ -233,7 +233,7 @@ config DAVINCI_WATCHDOG
 
 config ORION5X_WATCHDOG
 	tristate "Orion5x watchdog"
-	depends on ARCH_ORION5X
+	depends on ARCH_ORION5X && BROKEN
 	help
 	  Say Y here if to include support for the watchdog timer
 	  in the Orion5x ARM SoCs.
