diff -Naur a/drivers/net/Kconfig b/drivers/net/Kconfig
--- a/drivers/net/Kconfig	2007-05-21 22:04:46.000000000 -0600
+++ b/drivers/net/Kconfig	2007-05-21 21:06:52.000000000 -0600
@@ -2363,7 +2363,7 @@
 
 config CHELSIO_T3
 	tristate "Chelsio Communications T3 10Gb Ethernet support"
-	depends on PCI
+	depends on PCI && (!ARM || BROKEN)
 	select FW_LOADER
 	help
 	  This driver supports Chelsio T3-based gigabit and 10Gb Ethernet
@@ -2472,7 +2472,7 @@
 
 config NETXEN_NIC
 	tristate "NetXen Multi port (1/10) Gigabit Ethernet NIC"
-	depends on PCI && (!SPARC32 || BROKEN)
+	depends on PCI && (!SPARC32 || BROKEN) && (!ARM || BROKEN)
 	help
 	  This enables the support for NetXen's Gigabit Ethernet card.
 
diff -Naur a/drivers/net/wireless/bcm43xx/Kconfig b/drivers/net/wireless/bcm43xx/Kconfig
--- a/drivers/net/wireless/bcm43xx/Kconfig	2007-05-20 00:58:33.000000000 -0600
+++ b/drivers/net/wireless/bcm43xx/Kconfig	2007-05-21 21:56:15.000000000 -0600
@@ -1,6 +1,6 @@
 config BCM43XX
 	tristate "Broadcom BCM43xx wireless support"
-	depends on PCI && IEEE80211 && IEEE80211_SOFTMAC && WLAN_80211 && EXPERIMENTAL
+	depends on PCI && IEEE80211 && IEEE80211_SOFTMAC && WLAN_80211 && EXPERIMENTAL && (!ARM || BROKEN)
 	select WIRELESS_EXT
 	select FW_LOADER
 	select HW_RANDOM
diff -Naur a/drivers/media/video/bt8xx/Kconfig b/drivers/media/video/bt8xx/Kconfig
--- a/drivers/media/video/bt8xx/Kconfig	2007-04-25 21:08:32.000000000 -0600
+++ b/drivers/media/video/bt8xx/Kconfig	2007-05-28 18:12:26.000000000 -0600
@@ -1,6 +1,6 @@
 config VIDEO_BT848
 	tristate "BT848 Video For Linux"
-	depends on VIDEO_DEV && PCI && I2C && VIDEO_V4L1
+	depends on VIDEO_DEV && PCI && I2C && VIDEO_V4L1 && (!ARM || BROKEN)
 	select I2C_ALGOBIT
 	select FW_LOADER
 	select VIDEO_BTCX
