Fix the interrupt of the 2nd Ethernet slot.

--- a/arch/arm/mach-iop3xx/n2100.c~	2006-10-26 22:23:24.053694788 +0000
+++ b/arch/arm/mach-iop3xx/n2100.c	2006-10-26 22:23:41.955186517 +0000
@@ -85,7 +85,7 @@
 		irq = IRQ_IOP321_XINT0;
 	} else if (PCI_SLOT(dev->devfn) == 2) {
 		/* RTL8110SB #2 */
-		irq = IRQ_IOP321_XINT1;
+		irq = IRQ_IOP321_XINT3;
 	} else if (PCI_SLOT(dev->devfn) == 3) {
 		/* Sil3512 */
 		irq = IRQ_IOP321_XINT2;
