diff -Naur linux-2.6.18.orig/include/asm-arm/arch-ixp4xx/nslu2.h linux-2.6.18/include/asm-arm/arch-ixp4xx/nslu2.h
--- linux-2.6.18.orig/include/asm-arm/arch-ixp4xx/nslu2.h	2006-09-19 21:42:06.000000000 -0600
+++ linux-2.6.18/include/asm-arm/arch-ixp4xx/nslu2.h	2006-11-16 06:31:35.000000000 -0700
@@ -84,8 +84,8 @@
 #define NSLU2_LED_RED_BM	(1L << NSLU2_LED_RED)
 #define NSLU2_LED_GRN_BM	(1L << NSLU2_LED_GRN)
 
-#define NSLU2_LED_DISK1		NSLU2_GPIO2
-#define NSLU2_LED_DISK2		NSLU2_GPIO3
+#define NSLU2_LED_DISK1		NSLU2_GPIO3
+#define NSLU2_LED_DISK2		NSLU2_GPIO2
 
 #define NSLU2_LED_DISK1_BM	(1L << NSLU2_GPIO2)
 #define NSLU2_LED_DISK2_BM	(1L << NSLU2_GPIO3)
