diff -ruN a/drivers/usb/misc/Kconfig b/drivers/usb/misc/Kconfig
--- a/drivers/usb/misc/Kconfig	2006-11-29 21:57:37.000000000 +0000
+++ b/drivers/usb/misc/Kconfig	2007-01-05 15:22:14.000000000 +0000
@@ -4,34 +4,6 @@
 comment "USB Miscellaneous drivers"
 	depends on USB
 
-config USB_EMI62
-	tristate "EMI 6|2m USB Audio interface support"
-	depends on USB
-	---help---
-	  This driver loads firmware to Emagic EMI 6|2m low latency USB
-	  Audio and Midi interface.
-
-	  After firmware load the device is handled with standard linux
-	  USB Audio driver.
-
-	  This code is also available as a module ( = code which can be
-	  inserted in and removed from the running kernel whenever you want).
-	  The module will be called audio. If you want to compile it as a
-	  module, say M here and read <file:Documentation/kbuild/modules.txt>.
-
-config USB_EMI26
-	tristate "EMI 2|6 USB Audio interface support"
-	depends on USB
-	---help---
-	  This driver loads firmware to Emagic EMI 2|6 low latency USB
-	  Audio interface.
-
-	  After firmware load the device is handled with standard linux
-	  USB Audio driver.
-
-	  To compile this driver as a module, choose M here: the
-	  module will be called emi26.
-
 config USB_ADUTUX
 	tristate "ADU devices from Ontrak Control Systems (EXPERIMENTAL)"
 	depends on USB && EXPERIMENTAL
diff -ruN a/drivers/usb/misc/Makefile b/drivers/usb/misc/Makefile
--- a/drivers/usb/misc/Makefile	2006-11-29 21:57:37.000000000 +0000
+++ b/drivers/usb/misc/Makefile	2007-01-05 15:22:03.000000000 +0000
@@ -7,8 +7,6 @@
 obj-$(CONFIG_USB_AUERSWALD)	+= auerswald.o
 obj-$(CONFIG_USB_CYPRESS_CY7C63)+= cypress_cy7c63.o
 obj-$(CONFIG_USB_CYTHERM)	+= cytherm.o
-obj-$(CONFIG_USB_EMI26)		+= emi26.o
-obj-$(CONFIG_USB_EMI62)		+= emi62.o
 obj-$(CONFIG_USB_FTDI_ELAN)	+= ftdi-elan.o
 obj-$(CONFIG_USB_IDMOUSE)	+= idmouse.o
 obj-$(CONFIG_USB_LCD)		+= usblcd.o
