mark ip2 broken due to not using request_firmware()

diff --git a/drivers/char/Kconfig b/drivers/char/Kconfig
index 2d854bb..7d3f39e 100644
--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -108,6 +108,7 @@ config SERIAL_NONSTANDARD
 
 config COMPUTONE
 	tristate "Computone IntelliPort Plus serial support"
+	depends on BROKEN
 	depends on SERIAL_NONSTANDARD && (ISA || EISA || PCI)
 	---help---
 	  This driver supports the entire family of Intelliport II/Plus
