diff -urpN a/drivers/net/bnx2.c b/drivers/net/bnx2.c
--- a/drivers/net/bnx2.c	2009-03-30 00:56:18.000000000 -0600
+++ b/drivers/net/bnx2.c	2009-03-30 00:59:15.000000000 -0600
@@ -59,8 +59,8 @@
 #define PFX DRV_MODULE_NAME	": "
 #define DRV_MODULE_VERSION	"1.9.3"
 #define DRV_MODULE_RELDATE	"March 17, 2009"
-#define FW_FILE_06		"bnx2-06-4.4.1.fw"
-#define FW_FILE_09		"bnx2-09-4.4.26.fw"
+#define FW_FILE_06		"bnx2-06-4.6.16.fw"
+#define FW_FILE_09		"bnx2-09-4.6.15.fw"
 
 #define RUN_AT(x) (jiffies + (x))
 
