diff --git a/drivers/staging/wlags49_h25/Kconfig b/drivers/staging/wlags49_h25/Kconfig
index 304a8c9..24365a8 100644
--- a/drivers/staging/wlags49_h25/Kconfig
+++ b/drivers/staging/wlags49_h25/Kconfig
@@ -1,6 +1,7 @@
 config WLAGS49_H25
 	tristate "Linksys HERMES II.5 WCF54G_Wireless-G_CompactFlash_Card"
 	depends on WLAN_80211 && WIRELESS_EXT && PCMCIA
+	depends on BROKEN
 	select WEXT_SPY
 	---help---
 	  Driver for wireless cards using Agere's HERMES II.5 chipset
