--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -100,15 +99,0 @@
-config COMPUTONE
-	tristate "Computone IntelliPort Plus serial support"
-	depends on SERIAL_NONSTANDARD && (ISA || EISA || PCI)
-	---help---
-	  This driver supports the entire family of Intelliport II/Plus
-	  controllers with the exception of the MicroChannel controllers and
-	  products previous to the Intelliport II. These are multiport cards,
-	  which give you many serial ports. You would need something like this
-	  to connect more than two modems to your Linux box, for instance in
-	  order to become a dial-in server. If you have a card like that, say
-	  Y here and read <file:Documentation/computone.txt>.
-
-	  To compile this driver as modules, choose M here: the
-	  modules will be called ip2 and ip2main.
-
--- a/drivers/char/Makefile
+++ b/drivers/char/Makefile
@@ -36 +35,0 @@
-obj-$(CONFIG_COMPUTONE)		+= ip2/
