--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -2229,6 +2229,7 @@ config VIA_VELOCITY
 
 config TIGON3
 	tristate "Broadcom Tigon3 support"
+	depends on BROKEN
 	depends on PCI
 	help
 	  This driver supports Broadcom Tigon3 based gigabit Ethernet cards.
--- a/drivers/net/tg3.c
+++ b/drivers/net/tg3.c
@@ -5,14 +5,6 @@
  * Copyright (C) 2001, 2002, 2003 Jeff Garzik (jgarzik@pobox.com)
  * Copyright (C) 2004 Sun Microsystems Inc.
  * Copyright (C) 2005-2007 Broadcom Corporation.
- *
- * Firmware is:
- *	Derived from proprietary unpublished source code,
- *	Copyright (C) 2000-2003 Broadcom Corporation.
- *
- *	Permission is hereby granted for the distribution of this firmware
- *	data in hexadecimal or equivalent format, provided this copyright
- *	notice is accompanying it.
  */
 
 
