diff -ruN a/drivers/media/Kconfig b/drivers/media/Kconfig
--- a/drivers/media/Kconfig	2007-02-04 19:44:54.000000000 +0100
+++ b/drivers/media/Kconfig	2007-02-05 16:24:30.000000000 +0100
@@ -85,16 +85,0 @@
-if DAB
-config USB_DABUSB
-	tristate "DABUSB driver"
-	depends on USB
-	---help---
-	  A Digital Audio Broadcasting (DAB) Receiver for USB and Linux
-	  brought to you by the DAB-Team
-	  <http://wwwbode.cs.tum.edu/Par/arch/dab/>.  This driver can be taken
-	  as an example for URB-based bulk, control, and isochronous
-	  transactions. URB's are explained in
-	  <Documentation/usb/URB.txt>.
-
-	  To compile this driver as a module, choose M here: the
-	  module will be called dabusb.
-endif # DAB
-
diff -ruN a/drivers/media/video/Makefile b/drivers/media/video/Makefile
--- a/drivers/media/video/Makefile	2007-02-04 19:44:54.000000000 +0100
+++ b/drivers/media/video/Makefile	2007-02-05 16:24:30.000000000 +0100
@@ -94,1 +94,0 @@
-obj-$(CONFIG_USB_DABUSB)        += dabusb.o
