diff -ruN a/drivers/net/tokenring/Kconfig b/drivers/net/tokenring/Kconfig
--- a/drivers/net/tokenring/Kconfig	2006-11-29 21:57:37.000000000 +0000
+++ b/drivers/net/tokenring/Kconfig	2007-01-05 15:26:27.000000000 +0000
@@ -64,24 +64,6 @@
 	  To compile this driver as a module, choose M here: the module will be
 	  called lanstreamer.
 
-config 3C359
-	tristate "3Com 3C359 Token Link Velocity XL adapter support"
-	depends on TR && PCI
-	---help---
-	  This is support for the 3Com PCI Velocity XL cards, specifically
-	  the 3Com 3C359, please note this is not for the 3C339 cards, you
-	  should use the tms380 driver instead.
-
-	  If you have such an adapter, say Y and read the Token-Ring
-	  mini-HOWTO, available from <http://www.tldp.org/docs.html#howto>.
-
-	  To compile this driver as a module, choose M here: the module will be
-	  called 3c359.
-
-	  Also read the file <file:Documentation/networking/3c359.txt> or check the 
-	  Linux Token Ring Project site for the latest information at
-	  <http://www.linuxtr.net>
-
 config TMS380TR
 	tristate "Generic TMS380 Token Ring ISA/PCI adapter support"
 	depends on TR && (PCI || ISA && ISA_DMA_API || MCA)
@@ -166,21 +148,5 @@
 	  To compile this driver as a module, choose M here: the module will be
 	  called madgemc.
 
-config SMCTR
-	tristate "SMC ISA/MCA adapter support"
-	depends on TR && (ISA || MCA_LEGACY) && (BROKEN || !64BIT)
-	---help---
-	  This is support for the ISA and MCA SMC Token Ring cards,
-	  specifically SMC TokenCard Elite (8115T) and SMC TokenCard Elite/A
-	  (8115T/A) adapters.
-
-	  If you have such an adapter and would like to use it, say Y or M and
-	  read the Token-Ring mini-HOWTO, available from
-	  <http://www.tldp.org/docs.html#howto> and the file
-	  <file:Documentation/networking/smctr.txt>.
-
-	  To compile this driver as a module, choose M here: the module will be
-	  called smctr.
-
 endmenu
 
diff -ruN a/drivers/net/tokenring/Makefile b/drivers/net/tokenring/Makefile
--- a/drivers/net/tokenring/Makefile	2006-11-29 21:57:37.000000000 +0000
+++ b/drivers/net/tokenring/Makefile	2007-01-05 15:26:04.000000000 +0000
@@ -11,5 +11,3 @@
 obj-$(CONFIG_PROTEON) 	+= proteon.o
 obj-$(CONFIG_TMSPCI) 	+= tmspci.o
 obj-$(CONFIG_SKISA) 	+= skisa.o
-obj-$(CONFIG_SMCTR) 	+= smctr.o
-obj-$(CONFIG_3C359)	+= 3c359.o
