From dd99cd4f032495eff278c128980c5e5b9a7b2047 Mon Sep 17 00:00:00 2001
From: Ben Hutchings <ben@decadent.org.uk>
Date: Wed, 15 Oct 2008 02:43:20 +0100
Subject: [PATCH 13/24] tehuti: Mark as broken

---
 drivers/net/Kconfig |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/net/Kconfig b/drivers/net/Kconfig
index 64a39af..7fdc18f 100644
--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -2557,6 +2557,7 @@ config MLX4_DEBUG
 config TEHUTI
 	tristate "Tehuti Networks 10G Ethernet"
 	depends on PCI
+	depends on BROKEN
 	help
 	  Tehuti Networks 10G Ethernet NIC
 
-- 
1.6.1.3

