diff --git a/drivers/net/Kconfig b/drivers/net/Kconfig
index 5b9e17b..d6ba3a9 100644
--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -2229,6 +2229,7 @@ config VIA_VELOCITY
 
 config TIGON3
 	tristate "Broadcom Tigon3 support"
+	depends on BROKEN
 	depends on PCI
 	help
 	  This driver supports Broadcom Tigon3 based gigabit Ethernet cards.
diff --git a/drivers/net/tg3.c b/drivers/net/tg3.c
index 9034a05..a3ad51e 100644
--- a/drivers/net/tg3.c
+++ b/drivers/net/tg3.c
@@ -5131,6 +5131,7 @@ static int tg3_halt(struct tg3 *tp, int kind, int silent)
 	return 0;
 }
 
+#ifdef REMOVE_DFSG
 #define TG3_FW_RELEASE_MAJOR	0x0
 #define TG3_FW_RELASE_MINOR	0x0
 #define TG3_FW_RELEASE_FIX	0x0
@@ -5254,6 +5255,7 @@ u32 tg3FwData[(TG3_FW_DATA_LEN / sizeof(u32)) + 1] = {
 	0x00000000, 0x00000000, 0x00000000, 0x00000000
 };
 #endif
+#endif /* REMOVE_DFSG */
 
 #define RX_CPU_SCRATCH_BASE	0x30000
 #define RX_CPU_SCRATCH_SIZE	0x04000
@@ -5433,6 +5435,7 @@ static int tg3_load_5701_a0_firmware_fix(struct tg3 *tp)
 }
 
 
+#ifdef REMOVE_DFSG
 #define TG3_TSO_FW_RELEASE_MAJOR	0x1
 #define TG3_TSO_FW_RELASE_MINOR		0x6
 #define TG3_TSO_FW_RELEASE_FIX		0x0
@@ -5936,4 +5939,5 @@ static const u32 tg3Tso5FwData[(TG3_TSO5_FW_DATA_LEN / 4) + 1] = {
 	0x00000000, 0x00000000, 0x00000000,
 };
+#endif /* REMOVE_DFSG */
 
 /* tp->lock is held. */
