diff -ruN a/drivers/net/Kconfig b/drivers/net/Kconfig
--- a/drivers/net/Kconfig	2007-02-04 19:44:54.000000000 +0100
+++ b/drivers/net/Kconfig	2007-02-05 16:31:22.000000000 +0100
@@ -1471,15 +1471,0 @@
-config DGRS
-	tristate "Digi Intl. RightSwitch SE-X support"
-	depends on NET_PCI && (PCI || EISA)
-	---help---
-	  This is support for the Digi International RightSwitch series of
-	  PCI/EISA Ethernet switch cards. These include the SE-4 and the SE-6
-	  models.  If you have a network card of this type, say Y and read the
-	  Ethernet-HOWTO, available from
-	  <http://www.tldp.org/docs.html#howto>.  More specific
-	  information is contained in <file:Documentation/networking/dgrs.txt>.
-
-	  To compile this driver as a module, choose M here and read
-	  <file:Documentation/networking/net-modules.txt>.  The module
-	  will be called dgrs.
-
diff -ruN a/drivers/net/Makefile b/drivers/net/Makefile
--- a/drivers/net/Makefile	2007-02-04 19:44:54.000000000 +0100
+++ b/drivers/net/Makefile	2007-02-05 16:31:22.000000000 +0100
@@ -38,1 +38,0 @@
-obj-$(CONFIG_DGRS) += dgrs.o
