From e94e3972dc6c859ddca58baff52ea1c6c6953411 Mon Sep 17 00:00:00 2001
From: Ben Hutchings <ben@decadent.org.uk>
Date: Sat, 18 Oct 2008 14:07:10 +0100
Subject: [PATCH 19/24] typhoon: Mark as broken

---
 drivers/net/Kconfig |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/net/Kconfig b/drivers/net/Kconfig
index ded4b06..4bddcbd 100644
--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -732,6 +732,7 @@ config VORTEX
 config TYPHOON
 	tristate "3cr990 series \"Typhoon\" support"
 	depends on NET_VENDOR_3COM && PCI
+	depends on BROKEN
 	select CRC32
 	---help---
 	  This option enables driver support for the 3cr990 series of cards:
-- 
1.6.1.3

