diff --git a/drivers/staging/wlags49_h2/Kconfig b/drivers/staging/wlags49_h2/Kconfig
index 92053fe..b99b1b1 100644
--- a/drivers/staging/wlags49_h2/Kconfig
+++ b/drivers/staging/wlags49_h2/Kconfig
@@ -1,6 +1,7 @@
 config WLAGS49_H2
 	tristate "Agere Systems HERMES II Wireless PC Card Model 0110"
 	depends on WLAN_80211 && WIRELESS_EXT && PCMCIA
+	depends on BROKEN
 	select WEXT_SPY
 	---help---
 	  Driver for wireless cards using Agere's HERMES II chipset
