From: Ben Hutchings <ben@decadent.org.uk>
Date: Mon, 17 Aug 2009 02:17:09 +0100
Subject: [PATCH] ib_qib: disable use of SD7220

--- a/drivers/infiniband/hw/qib/Makefile
+++ b/drivers/infiniband/hw/qib/Makefile
@@ -6,7 +6,7 @@ ib_qib-y := qib_cq.o qib_diag.o qib_dma.o qib_driver.o qib_eeprom.o \
 	qib_qp.o qib_qsfp.o qib_rc.o qib_ruc.o qib_sdma.o qib_srq.o \
 	qib_sysfs.o qib_twsi.o qib_tx.o qib_uc.o qib_ud.o \
 	qib_user_pages.o qib_user_sdma.o qib_verbs_mcast.o qib_iba7220.o \
-	qib_sd7220.o qib_sd7220_img.o qib_iba7322.o qib_verbs.o
+	qib_sd7220.o qib_iba7322.o qib_verbs.o
 
 # 6120 has no fallback if no MSI interrupts, others can do INTx
 ib_qib-$(CONFIG_PCI_MSI) += qib_iba6120.o
--- a/drivers/infiniband/hw/qib/qib_sd7220.c
+++ b/drivers/infiniband/hw/qib/qib_sd7220.c
@@ -881,6 +902,18 @@ int qib_sd7220_prog_vfy(struct qib_devdata *dd, int sdnum,
 	return errors ? -errors : sofar;
 }
 
+int
+qib_sd7220_ib_load(struct qib_devdata *dd)
+{
+	return -1;
+}
+
+int
+qib_sd7220_ib_vfy(struct qib_devdata *dd)
+{
+	return -1;
+}
+
 /*
  * IRQ not set up at this point in init, so we poll.
  */
