diff --git a/drivers/media/video/Kconfig b/drivers/media/video/Kconfig
index ce1e7ba..4bd5a0c 100644
--- a/drivers/media/video/Kconfig
+++ b/drivers/media/video/Kconfig
@@ -483,6 +483,7 @@ config VIDEO_OV7670
 config VIDEO_VS6624
 	tristate "ST VS6624 sensor support"
 	depends on VIDEO_V4L2 && I2C
+	depends on BROKEN
 	---help---
 	  This is a Video4Linux2 sensor-level driver for the ST VS6624
 	  camera.
